<!DOCTYPE HTML>
<html lang="en" class="sidebar-visible no-js light">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>Modkit</title>
        <meta name="robots" content="noindex" />


        <!-- Custom HTML head -->
        
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff" />

        <link rel="icon" href="favicon.svg">
        <link rel="shortcut icon" href="favicon.png">
        <link rel="stylesheet" href="css/variables.css">
        <link rel="stylesheet" href="css/general.css">
        <link rel="stylesheet" href="css/chrome.css">
        <link rel="stylesheet" href="css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="fonts/fonts.css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="highlight.css">
        <link rel="stylesheet" href="tomorrow-night.css">
        <link rel="stylesheet" href="ayu-highlight.css">

        <!-- Custom theme stylesheets -->
        <link rel="stylesheet" href="custom.css">

        <!-- MathJax -->
        <script async src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    </head>
    <body>
    <div id="body-container">
        <!-- Provide site root to javascript -->
        <script>
            var path_to_root = "";
            var default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? "navy" : "light";
        </script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script>
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script>
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            var html = document.querySelector('html');
            html.classList.remove('no-js')
            html.classList.remove('light')
            html.classList.add(theme);
            html.classList.add('js');
        </script>

        <!-- Hide / unhide sidebar before it is displayed -->
        <script>
            var html = document.querySelector('html');
            var sidebar = null;
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            } else {
                sidebar = 'hidden';
            }
            html.classList.remove('sidebar-visible');
            html.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <div class="sidebar-scrollbox">
                <ol class="chapter"><li class="chapter-item expanded "><a href="quick_start.html"><strong aria-hidden="true">1.</strong> Quick Start guides</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="intro_bedmethyl.html"><strong aria-hidden="true">1.1.</strong> Constructing bedMethyl tables</a></li><li class="chapter-item expanded "><a href="intro_adjust.html"><strong aria-hidden="true">1.2.</strong> Updating and adjusting MM tags</a></li><li class="chapter-item expanded "><a href="intro_summary.html"><strong aria-hidden="true">1.3.</strong> Summarizing a modBAM</a></li><li class="chapter-item expanded "><a href="intro_motif_bed.html"><strong aria-hidden="true">1.4.</strong> Making a motif BED file</a></li><li class="chapter-item expanded "><a href="intro_extract.html"><strong aria-hidden="true">1.5.</strong> Extracting read information to a table</a></li><li class="chapter-item expanded "><a href="intro_call_mods.html"><strong aria-hidden="true">1.6.</strong> Calling mods in a modBAM</a></li><li class="chapter-item expanded "><a href="intro_edge_filter.html"><strong aria-hidden="true">1.7.</strong> Removing modification calls at the ends of reads</a></li><li class="chapter-item expanded "><a href="intro_include_bed.html"><strong aria-hidden="true">1.8.</strong> Narrow output to specific positions</a></li><li class="chapter-item expanded "><a href="intro_repair.html"><strong aria-hidden="true">1.9.</strong> Repair MM/ML tags on trimmed reads</a></li><li class="chapter-item expanded "><a href="intro_pileup_hemi.html"><strong aria-hidden="true">1.10.</strong> Make hemi-methylation bedMethyl tables</a></li><li class="chapter-item expanded "><a href="intro_dmr.html"><strong aria-hidden="true">1.11.</strong> Perform differential methylation scoring</a></li></ol></li><li class="chapter-item expanded "><a href="advanced_usage.html"><strong aria-hidden="true">2.</strong> Extended subcommand help</a></li><li class="chapter-item expanded "><a href="troubleshooting.html"><strong aria-hidden="true">3.</strong> Troubleshooting</a></li><li class="chapter-item expanded "><a href="limitations.html"><strong aria-hidden="true">4.</strong> Current limitations</a></li><li class="chapter-item expanded "><a href="perf_considerations.html"><strong aria-hidden="true">5.</strong> Performance considerations</a></li><li class="chapter-item expanded "><a href="algo_details.html"><strong aria-hidden="true">6.</strong> Algorithm details</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="filtering.html"><strong aria-hidden="true">6.1.</strong> Pass/fail base modification calls</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="filtering_details.html"><strong aria-hidden="true">6.1.1.</strong> Threshold examples</a></li><li class="chapter-item expanded "><a href="filtering_numeric_details.html"><strong aria-hidden="true">6.1.2.</strong> Numeric details</a></li></ol></li><li class="chapter-item expanded "><a href="collapse.html"><strong aria-hidden="true">6.2.</strong> Ignoring and combining calls</a></li></ol></li></ol>
            </div>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle"></div>
        </nav>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky bordered">
                    <div class="left-buttons">
                        <button id="sidebar-toggle" class="icon-button" type="button" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </button>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="light">Light</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">Modkit</h1>

                    <div class="right-buttons">
                        <a href="print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>

                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script>
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <h1 id="basic-usage"><a class="header" href="#basic-usage">Basic Usage</a></h1>
<h3 id="modkit-is-a-bioinformatics-tool-for-working-with-modified-bases-from-oxford-nanopore"><a class="header" href="#modkit-is-a-bioinformatics-tool-for-working-with-modified-bases-from-oxford-nanopore"><code>modkit</code> is a bioinformatics tool for working with modified bases from Oxford Nanopore.</a></h3>
<p><img src="./images/ONT_logo_590x106.png" alt="ONT_logo" /></p>
<h2 id="installation"><a class="header" href="#installation">Installation</a></h2>
<p>Pre-compiled binaries are provided for Linux from the <a href="https://github.com/nanoporetech/modkit/releases">release
page</a>. We recommend the use of these in
most circumstances. As a rust-based project, <code>modkit</code> can also be installed with 
<a href="https://www.rust-lang.org/learn/get-started">cargo</a>.</p>
<pre><code class="language-bash">git clone https://github.com/nanoporetech/modkit.git
cd modkit
cargo install --path .
# or
cargo install --git https://github.com/nanoporetech/modkit.git
</code></pre>
<h2 id="common-use-cases"><a class="header" href="#common-use-cases">Common Use Cases</a></h2>
<ol>
<li><a href="./intro_bedmethyl.html">Creating a bedMethyl table with <code>pileup</code></a></li>
<li><a href="./intro_adjust.html">Updating and Adjusting MM tags with <code>adjust-mods</code> and <code>update-tags</code></a></li>
<li><a href="./intro_summary.html">Summarizing a modBAM with <code>summary</code></a></li>
<li><a href="./intro_motif_bed.html">Making a motif BED file with <code>motif-bed</code></a></li>
<li><a href="./intro_extract.html">Extracting per-read base modification data into a table</a></li>
<li><a href="./intro_call_mods.html">Convert modification probabilities into hard calls</a></li>
<li><a href="./intro_edge_filter.html">Removing base modification calls at the ends of reads</a></li>
<li><a href="./intro_include_bed.html">Narrow analysis to only specific positions with a BED file</a></li>
<li><a href="./intro_repair.html">Repairing/adding MM/ML tags to reads with clipped sequences</a></li>
<li><a href="./intro_pileup_hemi.html">Creating hemi-methylation pattern bedMethyl tables with <code>pileup-hemi</code></a></li>
<li><a href="./intro_dmr.html">Performing differential methylation scoring with <code>dmr</code></a></li>
</ol>
<h2 id="notes-and-troubleshooting"><a class="header" href="#notes-and-troubleshooting">Notes and troubleshooting</a></h2>
<ol>
<li><a href="./troubleshooting.html">General troubleshooting</a></li>
<li><a href="./filtering_details.html">Threshold evaluation examples</a> (for advanced users)</li>
</ol>
<div style="break-before: page; page-break-before: always;"></div><h1 id="constructing-bedmethyl-tables"><a class="header" href="#constructing-bedmethyl-tables">Constructing bedMethyl tables.</a></h1>
<p>A primary use of <code>modkit</code> is to create summary counts of modified and unmodified bases in
an extended <a href="https://www.encodeproject.org/data-standards/wgbs/">bedMethyl</a> format.
bedMethyl files tabulate the counts of base modifications from every sequencing read over
each aligned reference genomic position. In order to create a bedMethyl table, your modBAM
must be aligned to a reference genome. The genome sequence is only required if you are using
the <code>--cpg</code> flag or <code>traditional</code> preset. Only <strong>primary alignments</strong> are used in generating 
the table, it is recommended to mark duplicate alignments before running as multiple primary
alignments can be double counted (but the behavior is logged). See <a href="./limitations.html">limitations</a>
for details.</p>
<h2 id="basic-usage-1"><a class="header" href="#basic-usage-1">Basic usage</a></h2>
<p>In its simplest form <code>modkit</code> creates a bedMethyl file using the following:</p>
<pre><code class="language-text">modkit pileup path/to/reads.bam output/path/pileup.bed --log-filepath pileup.log
</code></pre>
<p>No reference sequence is required. A single file (described
<a href="intro_bedmethyl.html#description-of-bedmethyl-output">below</a>) with base count summaries will be created. The
final argument here specifies an optional log file output.</p>
<p>The program performs best-practices filtering and manipulation of the raw data stored in
the input file. For further details see <a href="./filtering.html">filtering modified-base calls</a>.</p>
<h3 id="narrowing-output-to-cpg-dinucleotides"><a class="header" href="#narrowing-output-to-cpg-dinucleotides">Narrowing output to CpG dinucleotides</a></h3>
<p>For user convenience, the counting process can be modulated using several additional
transforms and filters. The most basic of these is to report only counts from reference
CpG dinucleotides. This option requires a reference sequence in order to locate the CpGs
in the reference:</p>
<pre><code class="language-bash">modkit pileup path/to/reads.bam output/path/pileup.bed --cpg --ref path/to/reference.fasta
</code></pre>
<p>To restrict output to only certain CpGs, pass the <code>--include-bed</code> option with the CpGs to be used, 
see <a href="./intro_include_bed.html">this page</a> for more details.</p>
<pre><code class="language-bash">modkit pileup path/to/reads.bam output/path/pileup.bed \
  --cpg \
  --ref path/to/reference.fasta \
  --include-bed path/to/my_cpgs.bed
</code></pre>
<p>The program also contains preset which combine several options for ease of use. The
<code>traditional</code> preset,</p>
<pre><code class="language-bash">modkit pileup path/to/reads.bam output/path/pileup.bed \
  --ref path/to/reference.fasta \
  --preset traditional
</code></pre>
<p>performs three transforms:</p>
<ul>
<li>restricts output to locations where there is a CG dinucleotide in the reference,</li>
<li>reports only a C and 5mC counts, using procedures to take into account counts of other
forms of cytosine modification (notably 5hmC), and</li>
<li>aggregates data across strands. The strand field of the output will be marked as '.'
indicating that the strand information has been lost.</li>
</ul>
<p>Using this option is equivalent to running with the options:</p>
<pre><code class="language-bash">modkit pileup path/to/reads.bam output/path/pileup.bed --cpg --ref &lt;reference.fasta&gt; --ignore h --combine-strands
</code></pre>
<h3 id="narrowing-output-to-specific-motifs"><a class="header" href="#narrowing-output-to-specific-motifs">Narrowing output to specific motifs</a></h3>
<p>By default, <code>modkit</code> will output a BED row for all genomic positions where
there is at least one base modification in the input modBAM. We define a motif
as a short DNA sequence potentially containing <a href="https://en.wikipedia.org/wiki/Nucleic_acid_notation">degenerate
codes</a>. To ease downstream
analysis, the <code>--motif &lt;Motif&gt; &lt;offset, 0-based&gt;</code> option can be used to
pre-filter and annotate the bedMethyl rows. The <code>--cpg</code> flag is a alias for
<code>--motif CG 0</code> where the sequence motif is <code>CG</code> and the offset is <code>0</code>, meaning
pileup base modification counts for the first <code>C</code> in the motif on the top
strand the second <code>C</code> (complement to <code>G</code>) on the bottom strand. Another example
may be <code>--motif GATC 1</code>, signaling to pileup counts for the <code>A</code> in the second
position on the top strand and the <code>A</code> in the third position on the bottom
strand.</p>
<p>When multiple motifs are specified the <code>name</code> column (<a href="intro_bedmethyl.html#bedmethyl-column-descriptions">column
4</a>), will indicate which motif the counts are
tabulated for. For example, if <code>--motif CGCG 2 --motif CG 0</code> are passed you may
see lines such as:</p>
<pre><code class="language-text">oligo_741_adapters  39 40 m,CG,0   4	-	39	40	255,0,0	4 100.00 4 0 0 0 0 0 0
oligo_741_adapters  39 40 m,CGCG,2 4	-	39	40	255,0,0	4 100.00 4 0 0 0 0 0 0

</code></pre>
<p>The <code>--combine-strands</code> flag can be combined with <code>--motif</code> however all motifs
must be reverse-complement palindromic (<code>CG</code> <em>is</em> a palindrome but <code>CHH</code> is
not).</p>
<h3 id="partitioning-reads-based-on-sam-tag-values"><a class="header" href="#partitioning-reads-based-on-sam-tag-values">Partitioning reads based on SAM tag values</a></h3>
<p>If have a modBAM with reads from different conditions are other SAM tag annotations (for example <code>RG</code> or <code>HP</code>) you 
can pass the <code>--partition-tag</code> option and <code>modkit</code> will output a separate bedMethyl with counts for only the reads 
with that tag value. For example, if you have haplotype-annotated reads with the <code>HP</code> tag, you could use a command
like the following:</p>
<pre><code class="language-bash">modkit pileup path/to/reads.bam output/directory/ --cpg --ref &lt;reference.fasta&gt; --partition-tag HP --prefix haplotyped
</code></pre>
<p>The output will be multiple files in placed in <code>output/directory/haplotyped_&lt;1|2|etc&gt;.bed</code>, multiple <code>--partition-tag</code>
options can be passed and the output files will correspond to the observed combinations of tags found in the modBAM. 
For example if <code>--partition-tag RG</code> and <code>--partition-tag HP</code> are passed:</p>
<pre><code class="language-bash">outdir/
  &lt;prefix&gt;_&lt;RG_value_1&gt;_&lt;HP_value_1&gt;.bed
  &lt;prefix&gt;_&lt;RG_value_2&gt;_&lt;HP_value_1&gt;.bed
  &lt;prefix&gt;_&lt;RG_value_1&gt;_&lt;HP_value_2&gt;.bed
  &lt;prefix&gt;_&lt;RG_value_2&gt;_&lt;HP_value_2&gt;.bed
  # ... etc
</code></pre>
<p>Note that only tag values that can be easily turned into strings will be considered valid (e.g. numbers, characters,
strings, etc.), array values will not be used, and will result in <code>missing</code> being used. Reads missing all of the 
SAM tags will be put in <code>ungrouped.bed</code>.</p>
<p>For more information on the individual options see the <a href="./advanced_usage.html">Advanced Usage</a> help document.</p>
<h2 id="description-of-bedmethyl-output"><a class="header" href="#description-of-bedmethyl-output">Description of bedMethyl output.</a></h2>
<p>Below is a description of the bedMethyl columns generated by <code>modkit pileup</code>. A brief description of the
bedMethyl specification can be found on <a href="https://www.encodeproject.org/data-standards/wgbs/">Encode</a>.</p>
<h3 id="definitions"><a class="header" href="#definitions">Definitions:</a></h3>
<ul>
<li>N<sub>mod</sub> - Number of calls passing filters that were classified as a residue with a specified base modification.</li>
<li>N<sub>canonical</sub> - Number of calls passing filters were classified as the canonical base rather than modified. The
exact base must be inferred by the modification code. For example, if the modification code is <code>m</code> (5mC) then
the canonical base is cytosine. If the modification code is <code>a</code>, the canonical base is adenine.</li>
<li>N<sub>other mod</sub> - Number of calls passing filters that were classified as modified, but where the modification is different from the listed base (and the corresponding canonical base is equal). For example, for a given cytosine there may be 3 reads with
<code>h</code> calls, 1 with a canonical call, and 2 with <code>m</code> calls. In the bedMethyl row for <code>h</code> N<sub>other_mod</sub> would be 2. In the
<code>m</code> row N<sub>other_mod</sub> would be 3.</li>
<li>N<sub>valid_cov</sub> - the valid coverage. N<sub>valid_cov</sub> = N<sub>mod</sub> + N<sub>other_mod</sub> + N<sub>canonical</sub>, also used as the <code>score</code> in the bedMethyl</li>
<li>N<sub>diff</sub> - Number of reads with a base other than the canonical base for this modification. For example, in a row
for <code>h</code> the canonical base is cytosine, if there are 2 reads with C-&gt;A substitutions, N<sub>diff</sub> will be 2.</li>
<li>N<sub>delete</sub> - Number of reads with a deletion at this reference position</li>
<li>N<sub>fail</sub> - Number of calls where the probability of the call was below the threshold. The threshold can be
set on the command line or computed from the data (usually failing the lowest 10th percentile of calls).</li>
<li>N<sub>nocall</sub> - Number of reads aligned to this reference position, with the correct canonical base, but without a base
modification call. This can happen, for example, if the model requires a CpG dinucleotide and the read has a
CG-&gt;CH substitution such that no modification call was produced by the basecaller.</li>
</ul>
<h3 id="bedmethyl-column-descriptions"><a class="header" href="#bedmethyl-column-descriptions">bedMethyl column descriptions.</a></h3>
<div class="table-wrapper"><table><thead><tr><th>column</th><th>name</th><th>description</th><th>type</th></tr></thead><tbody>
<tr><td>1</td><td>chrom</td><td>name of reference sequence from BAM header</td><td>str</td></tr>
<tr><td>2</td><td>start position</td><td>0-based start position</td><td>int</td></tr>
<tr><td>3</td><td>end position</td><td>0-based exclusive end position</td><td>int</td></tr>
<tr><td>4</td><td>modified base code and motif</td><td>single letter code for modified base and motif when more than one motif is used</td><td>str</td></tr>
<tr><td>5</td><td>score</td><td>Equal to N<sub>valid_cov</sub>.</td><td>int</td></tr>
<tr><td>6</td><td>strand</td><td>'+' for positive strand '-' for negative strand, '.' when strands are combined</td><td>str</td></tr>
<tr><td>7</td><td>start position</td><td>included for compatibility</td><td>int</td></tr>
<tr><td>8</td><td>end position</td><td>included for compatibility</td><td>int</td></tr>
<tr><td>9</td><td>color</td><td>included for compatibility, always 255,0,0</td><td>str</td></tr>
<tr><td>10</td><td>N<sub>valid_cov</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>11</td><td>fraction modified</td><td>N<sub>mod</sub> / N<sub>valid_cov</sub></td><td>float</td></tr>
<tr><td>12</td><td>N<sub>mod</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>13</td><td>N<sub>canonical</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>14</td><td>N<sub>other_mod</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>15</td><td>N<sub>delete</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>16</td><td>N<sub>fail</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>17</td><td>N<sub>diff</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>18</td><td>N<sub>nocall</sub></td><td>See definitions above.</td><td>int</td></tr>
</tbody></table>
</div>
<h2 id="performance-considerations"><a class="header" href="#performance-considerations">Performance considerations</a></h2>
<p>The <code>--interval-size</code>, <code>--threads</code>, <code>--chunk-size</code>, and <code>--max-depth</code> parameters can be used to tweak the parallelism and 
memory consumption of <code>modkit pileup</code>. The defaults should be suitable for most use cases, for more details see the
<a href="./advanced_usage.html">advanced usage</a> and <a href="./perf_considerations.html">performance considerations</a> sections.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="updating-and-adjusting-mm-tags"><a class="header" href="#updating-and-adjusting-mm-tags">Updating and Adjusting MM tags.</a></h1>
<p>The <code>adjust-mods</code> subcommand can be used to manipulate MM (and corresponding ML) tags in a
modBam. In general, these simple commands are run prior to <code>pileup</code>, visualization, or
other analysis. For <code>adjust-mods</code> and <code>update-tags</code>, if a correct <code>MN</code> tag is found, secondary and supplementary
alignments will be output. See <a href="./troubleshooting.html">troubleshooting</a> for details.</p>
<h2 id="ignoring-a-modification-class"><a class="header" href="#ignoring-a-modification-class">Ignoring a modification class.</a></h2>
<p>To remove a base modification class from a modBAM and produce a new modBAM, use the
<code>--ignore</code> option for <code>adjust-mods</code>.</p>
<pre><code>modkit adjust-mods input.bam output.adjust.bam --ignore &lt;mod_code_to_ignore&gt;
</code></pre>
<p>For example the command below will remove 5hmC calls, leaving just 5mC calls.</p>
<pre><code>modkit adjust-mods input.bam output.adjust.bam --ignore h
</code></pre>
<p>For technical details on the transformation see <a href="./collapse.html#removing-dna-base-modification-probabilities">Removing modification calls from
BAMs</a>.</p>
<h2 id="combining-base-modification-probabilities"><a class="header" href="#combining-base-modification-probabilities">Combining base modification probabilities.</a></h2>
<p>Combining base modification probabilities may be desirable for downstream analysis or
visualization. Unlike <code>--ignore</code> which removes the probability of a class, <code>--convert</code>
will sum the probability of one class with another if the second class already exists. For
example, the command below will convert probabilities associated with <code>h</code> probability into
<code>m</code> probability. If <code>m</code> already exists, the probabilities will be summed.  As described in
<a href="./intro_adjust.html#changing-the-base-modification-code">changing the modification code</a>,
if the second base modification code doesn't exist, the probabilities are left unchanged.</p>
<pre><code>modkit adjust-mods input.bam output.convert.bam --convert h m
</code></pre>
<h2 id="updating-the-flag--and-"><a class="header" href="#updating-the-flag--and-">Updating the flag (<code>?</code> and <code>.</code>).</a></h2>
<p>The <a href="https://samtools.github.io/hts-specs/SAMtags.pdf">specification</a> (Section 1.7) allows
for omission of the MM flag, however this may not be the intent of missing base
modification probabilities for some models. The command below will add or change the <code>?</code> flag to a modBAM.</p>
<pre><code>modkit adjust-mods input.bam output.bam --mode ambiguous
</code></pre>
<p>Another option is to set the flag to <code>.</code>, the &quot;implicitly canonical&quot; mode:</p>
<pre><code>modkit adjust-mods input.bam output.bam --mode implicit
</code></pre>
<h2 id="changing-the-base-modification-code"><a class="header" href="#changing-the-base-modification-code">Changing the base modification code.</a></h2>
<p>Some functions in <code>modkit</code> or other tools may require the mod-codes in the MM tag be in
the <a href="https://samtools.github.io/hts-specs/SAMtags.pdf">specification</a>. </p>
<p>For example, the following command will change <code>C+Z,</code> tags to <code>C+m,</code> tags.</p>
<pre><code>modkit adjust-mods input.bam output.bam --convert Z m
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="summarizing-a-modbam"><a class="header" href="#summarizing-a-modbam">Summarizing a modBAM.</a></h1>
<p>The <code>modkit summary</code> sub-command is intended for collecting read-level statistics on
either a sample of reads, a region, or an entire modBam.</p>
<h2 id="summarize-the-base-modification-calls-in-a-modbam"><a class="header" href="#summarize-the-base-modification-calls-in-a-modbam">Summarize the base modification calls in a modBAM.</a></h2>
<pre><code>modkit summary input.bam 
</code></pre>
<p>will output a table similar to this</p>
<pre><code class="language-bash">&gt; parsing region chr20  # only present if --region option is provided
&gt; sampling 10042 reads from BAM # modulated with --num-reads
&gt; calculating threshold at 10% percentile # modulated with --filter-percentile
&gt; calculated thresholds: C: 0.7167969 # calculated per-canonical base, on the fly
# bases             C
# total_reads_used  9989
# count_reads_C     9989
# pass_threshold_C  0.7167969
# region            chr20:0-64444167
 base  code  pass_count  pass_frac   all_count  all_frac
 C     m     1192533     0.58716166  1305956    0.5790408
 C     h     119937      0.0590528   195335     0.086608544
 C     -     718543      0.3537855   754087     0.33435062
</code></pre>
<h2 id="description-of-columns-in-modkit-summary"><a class="header" href="#description-of-columns-in-modkit-summary">Description of columns in <code>modkit summary</code>:</a></h2>
<h3 id="totals-table"><a class="header" href="#totals-table">Totals table</a></h3>
<p>The lines of the totals table are prefixed with a <code>#</code> character.</p>
<div class="table-wrapper"><table><thead><tr><th>row</th><th>name</th><th>description</th><th>type</th></tr></thead><tbody>
<tr><td>1</td><td>bases</td><td>comma-separated list of canonical bases with modification calls.</td><td>str</td></tr>
<tr><td>2</td><td>total_reads_used</td><td>total number of reads from which base modification calls were extracted</td><td>int</td></tr>
<tr><td>3+</td><td>count_reads_{base}</td><td>total number of reads that contained base modifications for {base}</td><td>int</td></tr>
<tr><td>4+</td><td>filter_threshold_{base}</td><td>filter threshold used for {base}</td><td>float</td></tr>
</tbody></table>
</div>
<h3 id="modification-calls-table"><a class="header" href="#modification-calls-table">Modification calls table</a></h3>
<p>The modification calls table follows immediately after the totals table.</p>
<div class="table-wrapper"><table><thead><tr><th>column</th><th>name</th><th>description</th><th>type</th></tr></thead><tbody>
<tr><td>1</td><td>base</td><td>canonical base with modification call</td><td>char</td></tr>
<tr><td>2</td><td>code</td><td>base modification code, or <code>-</code> for canonical</td><td>char</td></tr>
<tr><td>3</td><td>pass_count</td><td>total number of passing (confidence &gt;= threshold) calls for the modification in column 2</td><td>int</td></tr>
<tr><td>4</td><td>pass_frac</td><td>fraction of passing (&gt;= threshold) calls for the modification in column 2</td><td>float</td></tr>
<tr><td>5</td><td>all_count</td><td>total number of calls for the modification code in column 2</td><td>int</td></tr>
<tr><td>6</td><td>all_frac</td><td>fraction of all calls for the modification in column 2</td><td>float</td></tr>
</tbody></table>
</div>
<p>For more details on thresholds see <a href="./filtering.html">filtering base modification calls</a>.</p>
<p>By default <code>modkit summary</code> will only use ten thousand reads when generating the summary
(or fewer if the modBAM has fewer than that). To use all of the reads in the modBAM set
the <code>--no-sampling</code> flag.</p>
<pre><code>modkit summary input.bam --no-sampling
</code></pre>
<p>There are <code>--no-filtering</code>, <code>--filter-percentile</code>, and <code>--filter-threshold</code> options that
can be used with or without sampling.</p>
<h3 id="passing-a-threshold-directly"><a class="header" href="#passing-a-threshold-directly">Passing a threshold directly.</a></h3>
<p>To estimate the pass thresholds on a subset of reads, but then summarize <em>all</em> of the
reads, there is a two-step process. First, determine the thresholds with <code>modkit sample-probs</code> (see <a href="./advanced_usage.html#sample-probs">usage</a> for more details). Then run
<code>modkit summary</code> with the threshold value specified.</p>
<pre><code>modkit sample-probs input.bam [--sampling-frac &lt;frac&gt; | --num-reads &lt;num&gt;]
</code></pre>
<p>This command will output a table like this:</p>
<pre><code>&gt; sampling 10042 reads from BAM
 base  percentile  threshold
 C     10          0.6972656
 C     50          0.96484375
 C     90          0.9941406
</code></pre>
<p>You can then use pass this threshold directly to <code>modkit summary</code>:</p>
<pre><code>modkit summary input.bam \
    --filter-threshold 0.6972656 \ # filter 10% lowest confidence calls
    --no-sampling
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="making-a-motif-bed-file"><a class="header" href="#making-a-motif-bed-file">Making a motif BED file.</a></h1>
<p>Downstream analysis may require a <a href="https://en.wikipedia.org/wiki/BED_(file_format)">BED</a>
file to select motifs of interest. For example, selecting GATC motifs in <em>E. coli</em>. This
command requires a reference sequence in
<a href="https://en.wikipedia.org/wiki/FASTA_format">FASTA</a> a motif to find, which can include
<a href="https://en.wikipedia.org/wiki/Nucleic_acid_notation">IUPAC ambiguous bases</a> and a
position within the motif.</p>
<p>The following command would make a BED file for CG motifs.</p>
<pre><code>modkit motif-bed reference.fasta CG 0 1&gt; cg_modifs.bed
</code></pre>
<p>The output is directed to standard out.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="extracting-base-modification-information"><a class="header" href="#extracting-base-modification-information">Extracting base modification information</a></h1>
<p>The <code>modkit extract</code> sub-command will produce a table containing the base modification probabilities, 
the read sequence context, and optionally aligned reference information.
For <code>extract</code>, if a correct <code>MN</code> tag is found, secondary and supplementary alignments may be output with the <code>--allow-non-primary</code> flag. 
See <a href="./troubleshooting.html">troubleshooting</a> for details.</p>
<p>The table will by default contain unmapped sections of the read (soft-clipped sections, for example). 
To only include mapped bases use the <code>--mapped</code> flag. To only include sites of interest, pass a 
BED-formatted file to the <code>--include-bed</code> option. Similarly, to exclude sites, pass a BED-formatted
file to the <code>--exclude</code> option. One caution, the files generated by <code>modkit extract</code> can be large (2-2.5x
the size of the BAM). You may want to either use the <code>--num-reads</code> option, the <code>--region</code> option, or
pre-filter the modBAM ahead of time. You can also stream the output to stdout by setting the output to <code>-</code>
or <code>stdout</code> and filter the columns before writing to disk.</p>
<h2 id="description-of-output-table"><a class="header" href="#description-of-output-table">Description of output table</a></h2>
<div class="table-wrapper"><table><thead><tr><th>column</th><th>name</th><th>description</th><th>type</th></tr></thead><tbody>
<tr><td>1</td><td>read_id</td><td>name of the read</td><td>str</td></tr>
<tr><td>2</td><td>forward_read_position</td><td>0-based position on the forward-oriented read sequence</td><td>int</td></tr>
<tr><td>3</td><td>ref_position</td><td>aligned 0-based reference sequence position, -1 means unmapped</td><td>int</td></tr>
<tr><td>4</td><td>chrom</td><td>name of aligned contig, or '.' if the read is Gunmapped</td><td>str</td></tr>
<tr><td>5</td><td>mod_strand</td><td>strand of the molecule the base modification is on</td><td>str</td></tr>
<tr><td>6</td><td>ref_strand</td><td>strand of the reference the read is aligned to, or '.' if unmapped</td><td>str</td></tr>
<tr><td>7</td><td>ref_mod_strand</td><td>strand of the reference with the base modification, or '.' if unmapped</td><td>str</td></tr>
<tr><td>8</td><td>fw_soft_clipped_start</td><td>number of bases soft clipped from the start of the forward-oriented read</td><td>int</td></tr>
<tr><td>9</td><td>fw_soft_clipped_end</td><td>number of bases soft clipped from the end of the forward-oriented read</td><td>int</td></tr>
<tr><td>10</td><td>read_length</td><td>total length of the read</td><td>int</td></tr>
<tr><td>11</td><td>mod_qual</td><td>probability of the base modification in the next column</td><td>int</td></tr>
<tr><td>12</td><td>mod_code</td><td>base modification code from the MM tag</td><td>str</td></tr>
<tr><td>13</td><td>base_qual</td><td>basecall quality score (phred)</td><td>int</td></tr>
<tr><td>14</td><td>ref_kmer</td><td>reference 5-mer sequence context (center base is aligned base), '.' if unmapped</td><td>str</td></tr>
<tr><td>15</td><td>query_kmer</td><td>read 5-mer sequence context (center base is aligned base)</td><td>str</td></tr>
<tr><td>16</td><td>canonical_base</td><td>canonical base from the query sequence, from the MM tag</td><td>str</td></tr>
<tr><td>17</td><td>modified_primary_base</td><td>primary sequence base with the modification</td><td>str</td></tr>
<tr><td>18</td><td>inferred</td><td>whether the base modification call is implicit canonical</td><td>str</td></tr>
<tr><td>19</td><td>flag</td><td>FLAG from alignment record</td><td>str</td></tr>
</tbody></table>
</div>
<h1 id="tabulating-base-modification-calls-for-each-read-position"><a class="header" href="#tabulating-base-modification-calls-for-each-read-position">Tabulating base modification <em>calls</em> for each read position</a></h1>
<p>Passing <code>--read-calls &lt;file-path&gt;</code> option will generate a table of read-level base modification calls using the 
same <a href="./filtering.html">thresholding</a> algorithm employed by <code>modkit pileup</code>. The resultant table has, for each read,
one row for each base modification call in that read. If a base is called as modified then <code>call_code</code> will be the 
code in the <code>MM</code> tag. If the base is called as canonical the <code>call_code</code> will be <code>-</code> (<code>A</code>, <code>C</code>, <code>G</code>, and <code>T</code> are
reserved for &quot;any modification&quot;). The full schema of the table is below:</p>
<div class="table-wrapper"><table><thead><tr><th>column</th><th>name</th><th>description</th><th>type</th></tr></thead><tbody>
<tr><td>1</td><td>read_id</td><td>name of the read</td><td>str</td></tr>
<tr><td>2</td><td>forward_read_position</td><td>0-based position on the forward-oriented read sequence</td><td>int</td></tr>
<tr><td>3</td><td>ref_position</td><td>aligned 0-based reference sequence position, -1 means unmapped</td><td>int</td></tr>
<tr><td>4</td><td>chrom</td><td>name of aligned contig, or '.' if unmapped</td><td>str</td></tr>
<tr><td>5</td><td>mod_strand</td><td>strand of the molecule the base modification is on</td><td>str</td></tr>
<tr><td>6</td><td>ref_strand</td><td>strand of the reference the read is aligned to, or '.' if unmapped</td><td>str</td></tr>
<tr><td>7</td><td>ref_mod_strand</td><td>strand of the reference with the base modification, or '.' if unmapped</td><td>str</td></tr>
<tr><td>8</td><td>fw_soft_clipped_start</td><td>number of bases soft clipped from the start of the forward-oriented read</td><td>int</td></tr>
<tr><td>9</td><td>fw_soft_clipped_end</td><td>number of bases soft clipped from the end of the forward-oriented read</td><td>int</td></tr>
<tr><td>10</td><td>read_length</td><td>total length of the read</td><td>int</td></tr>
<tr><td>11</td><td>call_prob</td><td>probability of the base modification call in the next column</td><td>int</td></tr>
<tr><td>12</td><td>call_code</td><td>base modification call, <code>-</code> indicates a canonical call</td><td>str</td></tr>
<tr><td>13</td><td>base_qual</td><td>basecall quality score (phred)</td><td>int</td></tr>
<tr><td>14</td><td>ref_kmer</td><td>reference 5-mer sequence context (center base is aligned base), '.' if unmapped</td><td>str</td></tr>
<tr><td>15</td><td>query_kmer</td><td>read 5-mer sequence context (center base is aligned base)</td><td>str</td></tr>
<tr><td>16</td><td>canonical_base</td><td>canonical base from the query sequence, from the MM tag</td><td>str</td></tr>
<tr><td>17</td><td>modified_primary_base</td><td>primary sequence base with the modification</td><td>str</td></tr>
<tr><td>18</td><td>fail</td><td>true if the base modification call fell below the pass threshold</td><td>str</td></tr>
<tr><td>19</td><td>inferred</td><td>whether the base modification call is implicit canonical</td><td>str</td></tr>
<tr><td>20</td><td>within_alignment</td><td>when alignment information is present, is this base aligned to the reference</td><td>str</td></tr>
<tr><td>21</td><td>flag</td><td>FLAG from alignment record</td><td>str</td></tr>
</tbody></table>
</div>
<h2 id="note-on-implicit-base-modification-calls"><a class="header" href="#note-on-implicit-base-modification-calls">Note on implicit base modification calls.</a></h2>
<p>The <code>.</code> MM flag indicates that primary sequence bases without an associated base modification probability 
should be inferred to be canonical. By default, when this flag is encountered in a modBAM, <code>modkit extract</code> will 
output rows with the <code>inferred</code> column set to <code>true</code> and a <code>mod_qual</code> value of <code>0.0</code> for the base modifications
called on that read. For example, if you have a <code>A+a.</code> MM tag, and there are <code>A</code> bases in the read for which 
there aren't base modification calls (identifiable as non-0s in the MM tag) will be rows where the <code>mod_code</code> 
is <code>a</code> and the <code>mod_qual</code> is 0.0.</p>
<h2 id="note-on-non-primary-alignments"><a class="header" href="#note-on-non-primary-alignments">Note on non-primary alignments</a></h2>
<p>If a valid <code>MN</code> tag is found, secondary and supplementary alignments can be output in the <code>modkit extract</code> tables above.
See <a href="./troubleshooting.html">troubleshooting</a> for details on how to get valid <code>MN</code> tags.
To have non-primary alignments appear in the output, the <code>--allow-non-primary</code> flag must be passed. 
By default, the primary alignment will have all base modification information contained on the read, including soft-clipped and unaligned read positions. 
If the <code>--mapped-only</code> flag is used, soft clipped sections of the read will not be included. 
For secondary and supplementary alignments, soft-clipped positions are not repeated. See <a href="./advanced_usage.html">advanced usage</a> for more details.</p>
<h2 id="example-usages"><a class="header" href="#example-usages">Example usages:</a></h2>
<h3 id="extract-a-table-from-an-aligned-and-indexed-bam"><a class="header" href="#extract-a-table-from-an-aligned-and-indexed-bam">Extract a table from an aligned and indexed BAM</a></h3>
<pre><code>modkit extract &lt;input.bam&gt; &lt;output.tsv&gt; 
</code></pre>
<p>If the index <code>input.bam.bai</code> can be found, intervals along the aligned genome can be performed
in parallel.</p>
<h3 id="extract-a-table-from-a-region-of-a-large-modbam"><a class="header" href="#extract-a-table-from-a-region-of-a-large-modbam">Extract a table from a region of a large modBAM</a></h3>
<p>The below example will extract reads from only chr20, and include reference sequence context</p>
<pre><code>modkit extract &lt;intput.bam&gt; &lt;output.tsv&gt; --region chr20 --ref &lt;ref.fasta&gt;
</code></pre>
<h3 id="extract-only-sites-aligned-to-a-cg-motif"><a class="header" href="#extract-only-sites-aligned-to-a-cg-motif">Extract only sites aligned to a CG motif</a></h3>
<pre><code>modkit motif-bed &lt;reference.fasta&gt; CG 0 &gt; CG_motifs.bed
modkit extract &lt;in.bam&gt; &lt;out.tsv&gt; --ref &lt;ref.fasta&gt; --include-bed CG_motifs.bed
</code></pre>
<h3 id="extract-only-sites-that-are-at-least-50-bases-from-the-ends-of-the-reads"><a class="header" href="#extract-only-sites-that-are-at-least-50-bases-from-the-ends-of-the-reads">Extract only sites that are at least 50 bases from the ends of the reads</a></h3>
<pre><code>modkit extract &lt;in.bam&gt; &lt;out.tsv&gt; --edge-filter 50
</code></pre>
<h3 id="extract-read-level-base-modification-calls"><a class="header" href="#extract-read-level-base-modification-calls">Extract read-level base modification calls</a></h3>
<pre><code>modkit extract &lt;input.bam&gt; null --read-calls &lt;calls.tsv&gt;
</code></pre>
<p>Using &quot;null&quot; in the place of the normal output will direct the normal extract output
to /dev/null, to keep this output specify a file or <code>-</code> for standard out.</p>
<pre><code>modkit extract &lt;input.bam&gt; &lt;output.tsv&gt; --read-calls &lt;calls.tsv&gt;
</code></pre>
<p>Use <code>--allow-non-primary</code> to get secondary and supplementary mappings in the output.</p>
<pre><code>modkit extract &lt;input.bam&gt; &lt;output.tsv&gt; --read-calls &lt;calls.tsv&gt; --allow-non-primary
</code></pre>
<p>See the help string and/or <a href="./advanced_usage.html">advanced_usage</a> for more details.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="calling-mods-in-a-modbam"><a class="header" href="#calling-mods-in-a-modbam">Calling mods in a modBAM</a></h1>
<p>The <code>call-mods</code> subcommand in <code>modkit</code> transforms one modBAM into another
modBAM where the base modification probabilities have been clamped to 100% and
0%. If the <code>--filter-threshold</code> and/or <code>--mod-threshold</code>
<a href="./advanced_usage.html#call-mods">options</a> are provided, base modification calls
failing the threshold will be removed prior to changing the probabilities. The
output modBAM can be used for visualization, <code>pileup</code>, or other applications.
For <code>call-mods</code>, if a correct <code>MN</code> tag is found, secondary and supplementary
alignments will be output. See <a href="./troubleshooting.html">troubleshooting</a> for details.</p>
<p>A modBAM that has been transformed with <code>call-mods</code> using <code>--filter-threshold</code>
and/or <code>--mod-threshold</code> cannot be re-transformed with different thresholds.</p>
<p>Note on <code>pileup</code> with clamped probabilities: <code>modkit pileup</code> will attempt to
estimate the threshold probability by default, but it is unnecessary if the
modBAM is the result of <code>call-mods</code>. The threshold probabilities will be
artificially high (i.e. not representative of the model's output
probabilities). Similarly, specifying <code>--filter-threshold</code> and
<code>--mod-threshold</code> is not useful because all the ML probabilities have been set
to 0 and 100%.</p>
<h2 id="example-usages-1"><a class="header" href="#example-usages-1">Example usages</a></h2>
<h3 id="estimate-the-threshold-on-the-fly-apply-to-modbam-and-clamp-the-modification-calls-to-certainty"><a class="header" href="#estimate-the-threshold-on-the-fly-apply-to-modbam-and-clamp-the-modification-calls-to-certainty">Estimate the threshold on the fly, apply to modBAM and clamp the modification calls to certainty.</a></h3>
<pre><code>modkit call-mods &lt;in.bam&gt; &lt;out.bam&gt;
</code></pre>
<h3 id="specify-a-filter-threshold-for-your-use-case"><a class="header" href="#specify-a-filter-threshold-for-your-use-case">Specify a filter threshold for your use-case</a></h3>
<pre><code>modkit call-mods &lt;in.bam&gt; &lt;out.bam&gt; --filter-threshold A:0.9 --mod-threshold a:0.95 --filter-threshold C:0.97
</code></pre>
<h3 id="call-mods-with-the-estimated-threshold-and-ignore-modification-calls-within-100-base-pairs-of-the-ends-of-the-reds"><a class="header" href="#call-mods-with-the-estimated-threshold-and-ignore-modification-calls-within-100-base-pairs-of-the-ends-of-the-reds">Call mods with the estimated threshold and ignore modification calls within 100 base pairs of the ends of the reds</a></h3>
<pre><code>modkit call-mods &lt;in.bam&gt; &lt;out.bam&gt; --edge-filter 100
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="removing-modification-calls-at-the-ends-of-reads"><a class="header" href="#removing-modification-calls-at-the-ends-of-reads">Removing modification calls at the ends of reads</a></h1>
<p>If you have reads where you know base modifications near the ends should not be used 
(for example, if they are in adapters), you can use the <code>--edge-filter &lt;n_basepairs&gt;</code> option.
Two comma-separated values may be provided to asymmetrically filter out
base modification calls from the start and end of reads. For example, 4,8 will
filter out base modification calls in the first 4 and last 8 bases of the read. One value
will filter symmetrically. </p>
<ol>
<li><code>pileup</code>, will ignore base modification calls that are <code>&lt;n_basepairs&gt;</code> from the ends.</li>
<li><code>adjust-mods</code>, will <strong>remove</strong> base modification calls that are <code>&lt;n_basepairs&gt;</code> from the ends 
from the resultant output modBAM.</li>
<li><code>summary</code>, will ignore base modification calls that are <code>&lt;n_basepairs&gt;</code> from the ends.</li>
<li><code>sample-probs</code>, will ignore base modification calls that are <code>&lt;n_basepairs&gt;</code> from the ends.</li>
<li><code>call-mods</code>, will <strong>remove</strong> base modification calls that are <code>&lt;n_basepairs&gt;</code> from the ends from
the resultant output modBAM.</li>
<li><code>extract</code>, will ignore base modification calls that are <code>&lt;n_basepairs&gt;</code> from the ends, this also applies
when making the read-calls table (see <a href="./intro_extract.html">intro to extract</a>).</li>
</ol>
<p>In <code>pileup</code>, <code>call-mods</code>, and <code>extract</code> the edge-filter is also respected when estimating the pass-thresholds.
All commands have the flag <code>--invert-edge-filter</code> that will <em>keep</em> only base modification probabilities within
<code>&lt;n_basepairs&gt;</code> of the ends of the reads.</p>
<h2 id="example-usages-2"><a class="header" href="#example-usages-2">Example usages</a></h2>
<h3 id="call-mods-with-the-estimated-threshold-and-ignore-modification-calls-within-100-base-pairs-of-the-ends-of-the-reads"><a class="header" href="#call-mods-with-the-estimated-threshold-and-ignore-modification-calls-within-100-base-pairs-of-the-ends-of-the-reads">call mods with the estimated threshold and ignore modification calls within 100 base pairs of the ends of the reads</a></h3>
<pre><code>modkit call-mods &lt;in.bam&gt; &lt;out.bam&gt; --edge-filter 100
</code></pre>
<h3 id="perform-pileup-ignoring-base-modification-calls-within-100-base-pairs-of-the-ends-of-the-reads"><a class="header" href="#perform-pileup-ignoring-base-modification-calls-within-100-base-pairs-of-the-ends-of-the-reads">Perform pileup, ignoring base modification calls within 100 base pairs of the ends of the reads</a></h3>
<pre><code>modkit pileup &lt;in.bam&gt; &lt;out.bed&gt; --edge-filter 100
</code></pre>
<p>Filter out base modification calls within the first 25 bases or the last 10 bases.</p>
<pre><code>modkit pileup &lt;in.bam&gt; &lt;out.bed&gt; --edge-filter 25,10
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="narrow-output-to-specific-positions"><a class="header" href="#narrow-output-to-specific-positions">Narrow output to specific positions</a></h1>
<p>The <code>pileup</code>, <code>sample-probs</code>, <code>summary</code>, and <code>extract</code> sub commands have a <code>--include-bed</code> (or <code>--include-positions</code>) option that will restrict analysis to only positions that overlap with the intervals contained within the BED file. </p>
<p>In the case of <code>pileup</code>, <code>summary</code>, and <code>sample-probs</code>, the pass-threshold will be estimated with only base modification probabilities that are aligned to positions overlapping intervals in the BED. In the case of <code>pileup</code> and <code>extract</code> only positions will be reported if they overlap intervals in the BED.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="repair-mmml-tags-on-trimmed-reads"><a class="header" href="#repair-mmml-tags-on-trimmed-reads">Repair MM/ML tags on trimmed reads</a></h1>
<p>The <code>modkit repair</code> command is useful when you have a BAM with reads where the
canonical sequences have been altered in some way that either renders the <code>MM</code>
and <code>ML</code> tags invalid (for example, trimmed or hard-clipped) or the data has
been lost completely. This command requires that you have the original base
modification calls for each read you want to repair, and it will project these
base modification calls onto the sequences in the altered BAM.</p>
<p>The command uses two arguments called the &quot;donor&quot; and the &quot;acceptor&quot;. The
donor, contains the original, correct, <code>MM</code> and <code>ML</code> tags and the acceptor is
either missing <code>MM</code> and <code>ML</code> tags or they are invalid (they will be discarded
either way). The reads in the donor must be a superset of the reads in the
acceptor, meaning you can have extra reads in the donor BAM if some reads have
been removed or filtered earlier in the workflow. Both the donor and the
acceptor must be sorted by read name prior to running <code>modkit repair</code>.
Duplicate reads in the acceptor are allowed so long as they have valid SEQ
fields. Lastly, <code>modkit repair</code> only works on reads that have been trimmed,
other kinds of alteration such as run-length-encoding are not currently
supported. Split reads, or other derived transformations, are not currently
repairable with this command.</p>
<p>For example a typical workflow may look like this:</p>
<pre><code class="language-text"># original base modification calls
basecalls_5mC_5hmC.bam

# basecalls that have been trimmed
trimmed.bam # could also be fastq, but would require conversion to BAM

# the two BAM files need to be sorted
samtools -n trimmed.bam -O BAM &gt; trimed_read_sort.bam 
samtools -n basecalls_5mC_5hmC.bam -O BAM &gt; basecalls_5mC_5hmC_read_sort.bam

modkit repair \
    --donor-bam basecalls_5mC_5hmC_read_sort.bam \
    --acceptor-bam trimed_read_sort.bam \
    --log-filepath modkit_repair.log \
    --output-bam trimmed_repaired.bam
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="make-hemi-methylation-bedmethyl-tables-with-pileup-hemi"><a class="header" href="#make-hemi-methylation-bedmethyl-tables-with-pileup-hemi">Make hemi-methylation bedMethyl tables with <code>pileup-hemi</code></a></h1>
<p>Base modifications in DNA are inherently single-stranded, they (usually <sup class="footnote-reference"><a href="#1">1</a></sup>) don't change the base
pairing of the modified base. However, it may be of interest to know the correspondence
between the methylation state of a single base and another nearby base on the opposite strand - 
on the same molecule. In CpG dinucleotides, this is called &quot;hemi-methylation&quot;, when one cytosine
is methylated and the neighbor on the opposite strand is not:</p>
<pre><code class="language-text">     m
5'GATCGTACA
  CTAGCATGT
      -
</code></pre>
<p>In the above diagram, the cytosine in the fourth position on the positive strand is methylated (5mC) and the 
cytosine in the fifth position is canonical (-), indicating a &quot;hemi-methylation&quot;.</p>
<p>In the case of 5mC and canonical, there are 4 &quot;patterns&quot; of methylation:</p>
<pre><code class="language-text">m,m (5mC, 5mC)
-,m (canonical, 5mC)
m,- (5mC, canonical)
-,- (canonical, canonical)
</code></pre>
<p>These are all measured at the <em>single molecule</em> level, meaning each molecule must report on both strands (as
is the case with <a href="https://www.youtube.com/watch?v=8DVMG7FEBys">duplex</a> reads). For CpGs in the example above the
<code>MM</code> tags would be <code>C+m?</code> and <code>G-m?</code> for the top-strand and bottom-strand cytosines, respectively.</p>
<p>The <code>modkit pileup-hemi</code> command will perform an aggregation of the methylation &quot;patterns&quot; at genomic positions. An example
command to perform hemi-methylation analysis at CpGs would be</p>
<pre><code class="language-bash">modkit pileup-hemi \
  /path/to/duplex_reads.bam \
  --cpg \
  -r /path/to/reference.fasta \
  -o hemi_pileup.bed \
  --log modkit.log
</code></pre>
<p>Many of the <code>pileup</code> options are available in <code>pileup-hemi</code> with a couple differences: :</p>
<ol>
<li>A motif must be provided. The <code>--cpg</code> flag is a preset to aggregate CpG hemi-methylation patterns as shown above. 
If a motif is provided (as an argument to <code>--motif</code>) it must be reverse-complement palindromic.</li>
<li>A reference must be provided.</li>
<li>Both the positive strand base modification probability and the negative strand base modification probability must be above the pass threshold.</li>
</ol>
<p>See <a href="./advanced_usage.html">Advanced Usage</a> for details on all the options.</p>
<h2 id="description-of-hemi-methylation-patterns"><a class="header" href="#description-of-hemi-methylation-patterns">Description of hemi-methylation patterns</a></h2>
<p>The <code>modkit pileup-hemi</code> command aggregates a pair of base modification calls at each reference motif position
for each double-stranded DNA molecule. The base modification &quot;pattern&quot; indicates the methylation state on each base 
in 5-prime to 3-prime order, using the base modification code to indicate the identity of the base modification and 
<code>-</code> to indicate canonical (unmodified). For example <code>m,-,C</code> would mean the first base (from the reference 5' direction) 
is 5mC and the second base is unmodified and the primary base is cytosone. Similarly, <code>h,m,C</code> indicates the first base is 
5hmC and the second base is 5mC. The primary base called by the read is included to help disambiguate the unmodified
patterns (<code>-,-</code>). All patterns recognized at a location will be reported in the bedMethyl output.</p>
<h3 id="definitions-1"><a class="header" href="#definitions-1">Definitions:</a></h3>
<ul>
<li>N<sub>pattern</sub> - Number of call-pairs passing filters that had the pattern and primary base in column 4. E.g. <code>m,-,C</code> 
indicates the first base in the 5' to 3' direction is 5mC, the second base is unmodified and the primary base in the reads was C.</li>
<li>N<sub>canonical</sub> - Number of call-pairs passing filters that were classified as unmodified (i.e. the pattern is <code>-,-</code>).</li>
<li>N<sub>other_pattern</sub> - Number of call-pairs passing filters where the pattern is different from the pattern in 
column 4, but where the primary read base is the same. This count includes the unmodified pattern (<code>-,-</code>). <strong>Note</strong> this 
differs from <code>pileup</code> where N<sub>other</sub> does not contain the canonical counts.</li>
<li>N<sub>valid_cov</sub> - the valid coverage, total number of valid call-pairs.</li>
<li>N<sub>diff</sub> - Number of reads with a primary base other than the primary base in column 4.</li>
<li>N<sub>delete</sub> - Number of reads with a deletion at this reference position.</li>
<li>N<sub>fail</sub> - Number of call-pairs where the probability of the at least one of the calls in the pair was below 
the pass threshold. The threshold can be set on the command line or computed from the data (usually failing the 
lowest 10th percentile of calls).</li>
<li>N<sub>nocall</sub> - Number of reads where either one or both of the base modification calls was not present in the read.</li>
</ul>
<h3 id="bedmethyl-column-descriptions-1"><a class="header" href="#bedmethyl-column-descriptions-1">bedMethyl column descriptions.</a></h3>
<div class="table-wrapper"><table><thead><tr><th>column</th><th>name</th><th>description</th><th>type</th></tr></thead><tbody>
<tr><td>1</td><td>chrom</td><td>name of reference sequence from BAM header</td><td>str</td></tr>
<tr><td>2</td><td>start position</td><td>0-based start position</td><td>int</td></tr>
<tr><td>3</td><td>end position</td><td>0-based exclusive end position</td><td>int</td></tr>
<tr><td>4</td><td>methylation pattern</td><td>comma-separated pair of modification codes <code>-</code> means canonical, followed by the primary read base</td><td>str</td></tr>
<tr><td>5</td><td>score</td><td>Equal to N<sub>valid_cov</sub>.</td><td>int</td></tr>
<tr><td>6</td><td>strand</td><td>always '.' because strand information is combined</td><td>str</td></tr>
<tr><td>7</td><td>start position</td><td>included for compatibility</td><td>int</td></tr>
<tr><td>8</td><td>end position</td><td>included for compatibility</td><td>int</td></tr>
<tr><td>9</td><td>color</td><td>included for compatibility, always 255,0,0</td><td>str</td></tr>
<tr><td>10</td><td>N<sub>valid_cov</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>11</td><td>fraction modified</td><td>N<sub>pattern</sub> / N<sub>valid_cov</sub></td><td>float</td></tr>
<tr><td>12</td><td>N<sub>pattern</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>13</td><td>N<sub>canonical</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>14</td><td>N<sub>other_pattern</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>15</td><td>N<sub>delete</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>16</td><td>N<sub>fail</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>17</td><td>N<sub>diff</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>18</td><td>N<sub>nocall</sub></td><td>See definitions above.</td><td>int</td></tr>
</tbody></table>
</div>
<h2 id="limitations"><a class="header" href="#limitations">Limitations</a></h2>
<ol>
<li>Only one motif can be used at a time, this limitation may be removed in a later version.</li>
<li>Partitioning on tag key:value pairs is not currently supported.</li>
</ol>
<p><sup class="footnote-reference"><a href="#1">1</a></sup> In biology, there are almost always exceptions to every rule!</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="perform-differential-methylation-scoring"><a class="header" href="#perform-differential-methylation-scoring">Perform differential methylation scoring</a></h1>
<p>The <code>modkit dmr</code> command contains two subcommands, <code>pair</code> and <code>multi</code>, that will compare a pair
of samples and multiple samples, respectively. The details of <code>multi</code> are the same as <code>pair</code> (
it simply does all the pairwise comparisons), so most of the description below will focus on how
to run <code>pair</code> and how to interpret the outputs.</p>
<h2 id="preparing-the-input-data"><a class="header" href="#preparing-the-input-data">Preparing the input data</a></h2>
<p>The inputs to <code>modkit dmr</code> are two or more bedMethyl files (created by <code>modkit pileup</code>) that have
been compressed with <a href="https://www.htslib.org/doc/bgzip.html">bgzip</a> and indexed with 
<a href="https://www.htslib.org/doc/tabix.html">tabix</a>. An example workflow to generate the input data is shown below:</p>
<pre><code class="language-bash">ref=grch38.fasta
threads=32

norm=normal_sample.bam
norm_pileup=normal_pileup.bed

modkit pileup ${norm} ${norm_pileup} \
  --cpg \
  --ref ${ref} \
  --threads ${threads} \
  --log-filepath log.txt

bgzip ${norm_pileup}
tabix ${norm_pileup}.gz

tumor=tumor_sample.bam
tumor_pileup=tumor_pileup.bed

modkit pileup ${tumor} ${tumor_pileup} \
  --cpg \
  --ref ${ref} \
  --threads ${threads} \
  --log-filepath log.txt 

bgzip ${tumor_pileup}
tabix ${tumor_pileup}.gz
</code></pre>
<h2 id="running-differential-methylation-scoring"><a class="header" href="#running-differential-methylation-scoring">Running differential methylation scoring</a></h2>
<p>Once you have the two (or more) samples to be compared in the appropriate format, the final piece necessary 
is a BED file of the regions to be compared. The <code>modkit dmr</code> functionality does not &quot;segment&quot; or otherwise
discover regions, it scores the differences between user-provided regions. To continue with the above example
we can get CpG Islands from the <a href="http://genome.ucsc.edu/cgi-bin/hgTables">UCSC table browser</a>. The data may not 
always be appropriate input for <code>modkit</code>. For example, the CpG Islands track has extra columns and a header line:</p>
<pre><code class="language-text">#bin    chrom   chromStart      chromEnd        name       length  cpgNum  gcNum   perCpg  perGc   obsExp
1065    chr20   63004819        63007703        CpG: 272   2884    272     1869    18.9    64.8    0.9
1065    chr20   63009128        63009816        CpG: 48    688     48      432     14      62.8    0.71
</code></pre>
<p>Therefore, we need to transform the data with <code>awk</code> or similar, such as:</p>
<pre><code class="language-bash">awk 'BEGIN{FS=&quot;\t&quot;; OFS=&quot;\t&quot;} NR&gt;1 {print $2, $3, $4, $5}' cpg_islands_ucsc.bed \
  | bedtools sort -i - &gt;  cpg_islands_ucsc_cleaned.bed
</code></pre>
<p>Keeping the <code>name</code> column is optional. Sorting the regions isn't <em>strictly</em> necessary, the output will
be in the same order as the regions file. Below is an example command to produce the scored output
(continuing from the top example). The <code>--base</code> option tells <code>modkit dmr</code> which bases to use for scoring
the differences, the argument should be a canonical nucleotide (<code>A</code>, <code>C</code>, <code>G</code>, or <code>T</code>) whichever primary 
sequence base has the modifications you're interested in capturing. For example, for CpG islands the base
we're interested in is <code>C</code>.</p>
<pre><code class="language-bash">regions=cpg_islands_ucsc_cleaned.bed
dmr_result=cpg_islands_tumor_normal.bed

modkit dmr pair \
  -a ${norm_pileup}.gz \
  --index-a ${norm_pileup}.gz.tbi \ # optional
  -b ${tumor_pileup}.gz \
  --index-b ${tumor_pileup}.gz.tbi \ # optional
  -o ${dmr_result} \ # output to stdout if not present
  -r ${regions} \
  --ref ${ref} \
  --base C \  # may be repeated if multiple modifications are being used
  --threads ${threads} \
  --log-filepath dmr.log
</code></pre>
<h2 id="scoring-differentially-methylated-bases-as-opposed-to-regions"><a class="header" href="#scoring-differentially-methylated-bases-as-opposed-to-regions">Scoring differentially methylated bases (as opposed to regions)</a></h2>
<p>To score individual bases (e.g. differentially methylated CpGs), simply omit the <code>--regions</code> (<code>-r</code>) option
when running <code>modkit dmr [pair|multi]</code>. For example the above command becomes:</p>
<pre><code class="language-bash">dmr_result=cpg_islands_tumor_normal.bed

modkit dmr pair \
  -a ${norm_pileup}.gz \
  --index-a ${norm_pileup}.gz.tbi \ # optional
  -b ${tumor_pileup}.gz \
  --index-b ${tumor_pileup}.gz.tbi \ # optional
  -o ${dmr_result} \ # output to stdout if not present
  --ref ${ref} \
  --base C \  # may be repeated if multiple modifications are being used
  --threads ${threads} \
  --log-filepath dmr.log
</code></pre>
<h3 id="running-multiple-samples"><a class="header" href="#running-multiple-samples">Running multiple samples</a></h3>
<p>The <code>modkit dmr multi</code> command runs all pairwise comparisons for more than two samples.
The preparation of the data is identical to that for <code>dmr pair</code> (for each sample, of course). 
An example command could be:</p>
<pre><code class="language-bash">modkit dmr multi \
  -s ${norm_pileup_1}.gz norm1 \
  -s ${tumor_pileup_1}.gz tumor1 \
  -s ${norm_pileup_2}.gz norm2 \
  -s ${tumor_pileup_2}.gz tumor2 \
  -o ${dmr_dir} \ # required for multi
  -r ${cpg_islands} \ # skip this option to perform base-level DMR
  --ref ${ref} \
  --base C \
  -t 10 \
  -f \
  --log-filepath dmr_multi.log
</code></pre>
<p>For example the samples could be haplotype-partitioned bedMethyl tables or biological replicates.
Unlike for <code>modkit dmr pair</code> a sample name (e.g. <code>norm1</code> and <code>tumor1</code> above) must be provided for each input
sample. You can also use <code>--index &lt;filepath&gt; &lt;sample_name&gt;</code> to specify where the tabix index file is for each
sample.</p>
<h2 id="differential-methylation-output-format"><a class="header" href="#differential-methylation-output-format">Differential methylation output format</a></h2>
<p>The output from <code>modkit dmr pair</code> (and for each pairwise comparison with <code>modkit dmr multi</code>) is (roughly)
a BED file with the following schema:</p>
<div class="table-wrapper"><table><thead><tr><th>column</th><th>name</th><th>description</th><th>type</th></tr></thead><tbody>
<tr><td>1</td><td>chrom</td><td>name of reference sequence from bedMethyl input samples</td><td>str</td></tr>
<tr><td>2</td><td>start position</td><td>0-based start position, from <code>--regions</code> argument</td><td>int</td></tr>
<tr><td>3</td><td>end position</td><td>0-based exclusive end position, from <code>--regions</code> argument</td><td>int</td></tr>
<tr><td>4</td><td>name</td><td><code>name</code> column from <code>--regions</code> BED, or <code>chr:start-stop</code> if absent</td><td>str</td></tr>
<tr><td>5</td><td>score</td><td>Difference score, more positive values have increased difference</td><td>float</td></tr>
<tr><td>6</td><td>sample<sub>a</sub> counts</td><td>Counts of each base modification in the region, comma-separated, for sample A</td><td>str</td></tr>
<tr><td>7</td><td>sample<sub>a</sub> total</td><td>Total number of base modification calls in the region, including unmodified, for sample A</td><td>str</td></tr>
<tr><td>8</td><td>sample<sub>b</sub> counts</td><td>Counts of each base modification in the region, comma-separated, for sample B</td><td>str</td></tr>
<tr><td>9</td><td>sample<sub>b</sub> total</td><td>Total number of base modification calls in the region, including unmodified, for sample B</td><td>str</td></tr>
<tr><td>10</td><td>sample<sub>a</sub> fractions</td><td>Fraction of calls for each base modification in the region, comma-separated, for sample A</td><td>str</td></tr>
<tr><td>11</td><td>sample<sub>b</sub> fractions</td><td>Fraction of calls for each base modification in the region, comma-separated, for sample B</td><td>str</td></tr>
</tbody></table>
</div>
<p>an example of the output is given below:</p>
<pre><code class="language-text">chr10   73861   74083   chr10:73861-74083       -0.5007740865394226     h:7,m:18        950     h:8,m:16        802     h:0.74,m:1.89   h:1.00,m:2.00
chr10   74090   74289   chr10:74090-74289       0.5533780473006118      h:8,m:5         936     h:3,m:7         853     h:0.85,m:0.53   h:0.35,m:0.82
chr10   76139   76313   chr10:76139-76313       1.334274110255592       h:6,m:46        507     h:13,m:35       446     h:1.18,m:9.07   h:2.91,m:7.85
</code></pre>
<h2 id="scoring-details"><a class="header" href="#scoring-details">Scoring details</a></h2>
<p>The aim of <code>modkit dmr</code> is to enable exploratory data analysis of methylation patterns. To that aim, the approach to 
scoring methylation differences is intended to be simple and interpretable. For every region provided, within a sample, 
we model each potentially methylated base as arising from the same distribution. In other words, we discard the relative 
ordering of the base modification calls within a region. We then define a model for the frequency of observing each base 
modification state. In the case of methylated versus unmodified (5mC vs C, or 6mA vs A), we use the binomial distribution
and model the probability of methylation \(p\) as a beta-distributed random variable: </p>
<p>\[
\mathbf{X}|p \sim \text{Bin}(n, p)
\]
\[
p \sim \text{Beta}(\alpha, \beta)
\]</p>
<p>where \(n\) is the number of potentially methylated bases reported on in the 
region and \(\mathbf{X}\) is the vector of counts (canonical and methylated). </p>
<p>In the case where there are more than two states (for example, 5hmC, 5mC, and unmodified C) we use a multinomial 
distribution and a Dirichlet as the base distribution: 
\[
\mathbf{X}|\pi \sim \text{Mult}(n, \pi)
\]</p>
<p>\[
\pi \sim \text{Dir}(\alpha)
\]</p>
<p>Let \(\theta\) be the parameters describing the posterior distribution ( \( \alpha, \beta \) for the binary case, 
and \(\alpha \) in the general case). The <code>score</code> reported is the result of the following log marginal likelihood 
ratio :</p>
<p>\[
\text{score} = \text{log}(\frac{l_{\theta_{a}}( \mathbf{X_a} ) l_{\theta_{b}} ( \mathbf{X_b} )}{l_{\theta_{a+b}} (\mathbf{X_{a+b}} )})
\]</p>
<p>Where \(\theta_a\) and \(\theta_b\) are the posterior distributions with the two conditions modeled separately, 
and \(\theta_{a+b}\) is the posterior when the two conditions are modeled together. The function \(l_{\theta}(\mathbf{X}) \) is 
the log marginal likelihood of the counts under the parameters of the model \(\theta\).
For all cases, we use <a href="https://en.wikipedia.org/wiki/Jeffreys_prior">Jeffrey's prior</a> as the prior distribution.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="modkit-subcommand-documentation"><a class="header" href="#modkit-subcommand-documentation">modkit, subcommand documentation</a></h1>
<p>The goal of <code>modkit</code> is to enable best-practices manipulation of BAM files containing
modified base information (modBAMs). The various sub-commands and tools available in
<code>modkit</code> are described below.  This information can be obtained by invoking the long help
(<code>--help</code>) for each command.</p>
<blockquote>
<p>Advanced usage information.</p>
</blockquote>
<pre><code class="language-text">Usage: modkit &lt;COMMAND&gt;

Commands:
  pileup        Tabulates base modification calls across genomic positions. This command
                    produces a bedMethyl formatted file. Schema and description of fields can be
                    found in the README.
  adjust-mods   Performs various operations on BAM files containing base modification
                    information, such as converting base modification codes and ignoring
                    modification calls. Produces a BAM output file.
  update-tags   Renames Mm/Ml to tags to MM/ML. Also allows changing the the mode flag from
                    silent '.' to explicitly '?' or '.'.
  sample-probs  Calculate an estimate of the base modification probability distribution.
  summary       Summarize the mod tags present in a BAM and get basic statistics. The default
                    output is a totals table (designated by '#' lines) and a modification calls
                    table. Descriptions of the columns can be found in the README.
  call-mods     Call mods from a modbam, creates a new modbam with probabilities set to 100% if
                    a base modification is called or 0% if called canonical.
  motif-bed     Create BED file with all locations of a sequence motif. 
                    Example: modkit motif-bed CG 0
  extract       Extract read-level base modification information from a modBAM into a
                    tab-separated values table.
  repair        Repair MM and ML tags in one bam with the correct tags from another. To use this
                    command, both modBAMs _must_ be sorted by read name. The &quot;donor&quot; modBAM's reads
                    must be a superset of the acceptor's reads. Extra reads in the donor are
                    allowed, and multiple reads with the same name (secondary, etc.) are allowed in
                    the acceptor. Reads with an empty SEQ field cannot be repaired and will be
                    rejected. Reads where there is an ambiguous alignment of the acceptor to the
                    donor will be rejected (and logged). See the full documentation for details.
  dmr           Perform DMR test on a set of regions. Output a BED file of regions with the
                    score column indicating the magnitude of the difference. Find the schema and
                    description of fields can in the README as well as a description of the model
                    and method. See subcommand help for additional details.
  pileup-hemi   Tabulates double-stranded base modification patters (such as hemi-methylation)
                    across genomic motif positions. This command produces a bedMethyl file, the
                    schema can be found in the online documentation.
  help          Print this message or the help of the given subcommand(s).

Options:
  -h, --help     Print help information.
  -V, --version  Print version information.
</code></pre>
<h2 id="pileup"><a class="header" href="#pileup">pileup</a></h2>
<pre><code class="language-text">Tabulates base modification calls across genomic positions. This command produces a bedMethyl
formatted file. Schema and description of fields can be found in the README.

Usage: modkit pileup [OPTIONS] &lt;IN_BAM&gt; &lt;OUT_BED&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input BAM, should be sorted and have associated index available.

  &lt;OUT_BED&gt;
          Output file (or directory with --bedgraph option) to write results into. Specify &quot;-&quot; or
          &quot;stdout&quot; to direct output to stdout.

Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them. Setting a file is
          recommended. (alias: log)

      --region &lt;REGION&gt;
          Process only the specified region of the BAM when performing pileup. Format should be
          &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;. Commas are allowed.

      --max-depth &lt;MAX_DEPTH&gt;
          Maximum number of records to use when calculating pileup. This argument is passed to the
          pileup engine. If you have high depth data, consider increasing this value substantially.
          Must be less than 2147483647 or an error will be raised.
          
          [default: 8000]

  -t, --threads &lt;THREADS&gt;
          Number of threads to use while processing chunks concurrently.
          
          [default: 4]

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently. Smaller interval chunk sizes
          will use less memory but incur more overhead.
          
          [default: 100000]

      --chunk-size &lt;CHUNK_SIZE&gt;
          Break contigs into chunks containing this many intervals (see `interval_size`). This
          option can be used to help prevent excessive memory usage, usually with no performance
          penalty. By default, modkit will set this value to 1.5x the number of threads specified,
          so if 4 threads are specified the chunk_size will be 6. A warning will be shown if this
          option is less than the number of threads specified.

      --suppress-progress
          Hide the progress bar.

  -n, --num-reads &lt;NUM_READS&gt;
          Sample this many reads when estimating the filtering threshold. Reads will be sampled
          evenly across aligned genome. If a region is specified, either with the --region option or
          the --sample-region option, then reads will be sampled evenly across the region given.
          This option is useful for large BAM files. In practice, 10-50 thousand reads is sufficient
          to estimate the model output distribution and determine the filtering threshold.
          
          [default: 10042]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Sample this fraction of the reads when estimating the pass-threshold. In practice, 10-100
          thousand reads is sufficient to estimate the model output distribution and determine the
          filtering threshold. See filtering.md for details on filtering.

      --seed &lt;SEED&gt;
          Set a random seed for deterministic running, the default is non-deterministic.

      --no-filtering
          Do not perform any filtering, include all mod base calls in output. See filtering.md for
          details on filtering.

  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted variant is below
          this confidence percentile. For example, 0.1 will filter out the 10% lowest confidence
          modification calls.
          
          [default: 0.1]

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or per-base. Global filter threshold can be
          specified with by a decimal number (e.g. 0.75). Per-base thresholds can be specified by
          colon-separated values, for example C:0.75 specifies a threshold value of 0.75 for
          cytosine modification calls. Additional per-base thresholds can be specified by repeating
          the option: for example --filter-threshold C:0.75 --filter-threshold A:0.70 or specify a
          single base option and a default for all other bases with: --filter-threshold A:0.70
          --filter-threshold 0.9 will specify a threshold value of 0.70 for adenine and 0.9 for all
          other base modification calls.

      --mod-thresholds &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification, independent of the threshold
          for the primary sequence base or the default. For example, to set the pass threshold for
          5hmC to 0.8 use `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless the
          `--filter-threshold` option is also passed. See the online documentation for more details.

      --sample-region &lt;SAMPLE_REGION&gt;
          Specify a region for sampling reads from when estimating the threshold probability. If
          this option is not provided, but --region is provided, the genomic interval passed to
          --region will be used. Format should be &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;.

      --sampling-interval-size &lt;SAMPLING_INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently when estimating the threshold
          probability, can be larger than the pileup processing interval.
          
          [default: 1000000]

      --include-bed &lt;INCLUDE_BED&gt;
          BED file that will restrict threshold estimation and pileup results to positions
          overlapping intervals in the file. (alias: include-positions)

      --include-unmapped
          Include unmapped base modifications when estimating the pass threshold.

      --ignore &lt;IGNORE&gt;
          Ignore a modified base class  _in_situ_ by redistributing base modification probability
          equally across other options. For example, if collapsing 'h', with 'm' and canonical
          options, half of the probability of 'h' will be added to both 'm' and 'C'. A full
          description of the methods can be found in collapse.md.

      --force-allow-implicit
          Force allow implicit-canonical mode. By default modkit does not allow pileup with the
          implicit mode (e.g. C+m, no '.' or '?'). The `update-tags` subcommand is provided to
          update tags to the new mode. This option allows the interpretation of implicit mode tags:
          residues without modified base probability will be interpreted as being the non-modified
          base.

      --motif &lt;MOTIF&gt; &lt;MOTIF&gt;
          Output pileup counts for only sequence motifs provided. The first argument should be the
          sequence motif and the second argument is the 0-based offset to the base to pileup base
          modification counts for. For example: --motif CGCG 0 indicates to pileup counts for the
          first C on the top strand and the last C (complement to G) on the bottom strand. The --cpg
          argument is short hand for --motif CG 0.
          
          This argument can be passed multiple times. When more than one motif is used, the
          resulting output BED file will indicate the motif in the &quot;name&quot; field as
          &lt;mod_code&gt;,&lt;motif&gt;,&lt;offset&gt;. For example, given `--motif CGCG 2 --motif CG 0` there will
          be output lines with name fields such as &quot;m,CG,0&quot; and &quot;m,CGCG,2&quot;. To use this option with
          `--combine-strands`, all motifs must be reverse-complement palindromic or an error will be
          raised.

      --cpg
          Only output counts at CpG motifs. Requires a reference sequence to be provided.

  -r, --ref &lt;REFERENCE_FASTA&gt;
          Reference sequence in FASTA format. Required for CpG motif filtering.

  -k, --mask
          Respect soft masking in the reference FASTA.

      --preset &lt;PRESET&gt;
          Optional preset options for specific applications. traditional: Prepares bedMethyl
          analogous to that generated from other technologies for the analysis of 5mC modified
          bases. Shorthand for --cpg --combine-strands --ignore h.
          
          [possible values: traditional]

      --combine-mods
          Combine base modification calls, all counts of modified bases are summed together. See
          collapse.md for details.

      --combine-strands
          When performing motif analysis (such as CpG), sum the counts from the positive and
          negative strands into the counts for the positive strand position.

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the start or the end of the
          read. Two comma-separated values may be provided to asymmetrically filter out base
          modification calls from the start and end of the reads. For example, 4,8 will filter out
          base modification calls in the first 4 and last 8 bases of the read.

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification calls at the ends of
          reads, only _keep_ base modification calls at the ends of reads. E.g. if usually, &quot;4,8&quot;
          would remove (i.e. filter out) base modification calls in the first 4 and last 8 bases of
          the read, using this flag will keep only base modification calls in the first 4 and last 8
          bases.

      --only-tabs
          For bedMethyl output, separate columns with only tabs. The default is to use tabs for the
          first 10 fields and spaces thereafter. The default behavior is more likely to be
          compatible with genome viewers. Enabling this option may make it easier to parse the
          output with tabular data handlers that expect a single kind of separator.

      --bedgraph
          Output bedGraph format, see https://genome.ucsc.edu/goldenPath/help/bedgraph.html. For
          this setting, specify a directory for output files to be make in. Two files for each
          modification will be produced, one for the positive strand and one for the negative
          strand. So for 5mC (m) and 5hmC (h) there will be 4 files produced.

      --prefix &lt;PREFIX&gt;
          Prefix to prepend on bedgraph output file names. Without this option the files will be
          &lt;mod_code&gt;_&lt;strand&gt;.bedgraph.

      --partition-tag &lt;PARTITION_TAG&gt;
          Partition output into multiple bedMethyl files based on tag-value pairs. The output will
          be multiple bedMethyl files with the format
          `&lt;prefix&gt;_&lt;tag_value_1&gt;_&lt;tag_value_2&gt;_&lt;tag_value_n&gt;.bed` prefix is optional and set with
          the `--prefix` flag.

  -h, --help
          Print help information (use `-h` for a summary)
</code></pre>
<h2 id="adjust-mods"><a class="header" href="#adjust-mods">adjust-mods</a></h2>
<pre><code class="language-text">Performs various operations on BAM files containing base modification information, such as
converting base modification codes and ignoring modification calls. Produces a BAM output file

Usage: modkit adjust-mods [OPTIONS] &lt;IN_BAM&gt; &lt;OUT_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          BAM file to collapse mod call from. Can be a path to a file or one of `-` or `stdin` to
          specify a stream from standard input.

  &lt;OUT_BAM&gt;
          File path to new BAM file to be created. Can be a path to a file or one of `-` or `stdin`
          to specify a stream from standard output.

Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Output debug logs to file at this path.

      --ignore &lt;IGNORE&gt;
          Modified base code to ignore/remove, see https://samtools.github.io/hts-specs/SAMtags.pdf
          for details on the modified base codes.

  -t, --threads &lt;THREADS&gt;
          Number of threads to use.
          
          [default: 4]

  -f, --ff
          Fast fail, stop processing at the first invalid sequence record. Default behavior is to
          continue and report failed/skipped records at the end.

      --convert &lt;CONVERT&gt; &lt;CONVERT&gt;
          Convert one mod-tag to another, summing the probabilities together if the retained mod tag
          is already present.

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the start or the end of the
          read. Two comma-separated values may be provided to asymmetrically filter out base
          modification calls from the start and end of the reads. For example, 4,8 will filter out
          base modification calls in the first 4 and last 8 bases of the read.

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification calls at the ends of
          reads, only _keep_ base modification calls at the ends of reads. E.g. if usually, &quot;4,8&quot;
          would remove (i.e. filter out) base modification calls in the first 4 and last 8 bases of
          the read, using this flag will keep only base modification calls in the first 4 and last 8
          bases.

      --output-sam
          Output SAM format instead of BAM.

      --suppress-progress
          Hide the progress bar

  -h, --help
          Print help information (use `-h` for a summary).
</code></pre>
<h2 id="update-tags"><a class="header" href="#update-tags">update-tags</a></h2>
<pre><code class="language-text">Renames Mm/Ml to tags to MM/ML. Also allows changing the the mode flag from silent '.' to explicitly
'?' or '.'.

Usage: modkit update-tags [OPTIONS] &lt;IN_BAM&gt; &lt;OUT_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;   BAM to update modified base tags in. Can be a path to a file or one of `-` or `stdin`
             to specify a stream from standard input.
  &lt;OUT_BAM&gt;  File to new BAM file to be created or one of `-` or `stdin` to specify a stream from
             standard output.

Options:
  -m, --mode &lt;MODE&gt;                  Mode, change mode to this value, options {'ambiguous',
                                     'implicit'}. See spec at:
                                     https://samtools.github.io/hts-specs/SAMtags.pdf. 'ambiguous'
                                     ('?') means residues without explicit modification
                                     probabilities will not be assumed canonical or modified.
                                     'implicit' means residues without explicit modification
                                     probabilities are assumed to be canonical. [possible values:
                                     ambiguous, implicit]
  -t, --threads &lt;THREADS&gt;            Number of threads to use. [default: 4]
      --log-filepath &lt;LOG_FILEPATH&gt;  Output debug logs to file at this path.
      --output-sam                   Output SAM format instead of BAM.
  -h, --help                         Print help information.
</code></pre>
<h2 id="sample-probs"><a class="header" href="#sample-probs">sample-probs</a></h2>
<pre><code class="language-text">Calculate an estimate of the base modification probability distribution.

Usage: modkit sample-probs [OPTIONS] &lt;IN_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input BAM with modified base tags. If a index is found reads will be sampled evenly across
          the length of the reference sequence. Can be a path to a file or one of `-` or `stdin` to
          specify a stream from standard input.

Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use.
          
          [default: 4]

      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them. Setting a file is
          recommended.

      --suppress-progress
          Hide the progress bar.

  -p, --percentiles &lt;PERCENTILES&gt;
          Percentiles to calculate, a space separated list of floats.
          
          [default: 0.1,0.5,0.9]

  -o, --out-dir &lt;OUT_DIR&gt;
          Directory to deposit result tables into. Required for model probability histogram output.
          Creates two files probabilities.tsv and probabilities.txt The .txt contains
          ASCII-histograms and the .tsv contains tab-separated variable data represented by the
          histograms.

      --prefix &lt;PREFIX&gt;
          Label to prefix output files with. E.g. 'foo' will output foo_thresholds.tsv,
          foo_probabilities.tsv, and foo_probabilities.txt.

      --force
          Overwrite results if present.

      --ignore &lt;IGNORE&gt;
          Ignore a modified base class  _in_situ_ by redistributing base modification probability
          equally across other options. For example, if collapsing 'h', with 'm' and canonical
          options, half of the probability of 'h' will be added to both 'm' and 'C'. A full
          description of the methods can be found in collapse.md.

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the start or the end of the
          read. Two comma-separated values may be provided to asymmetrically filter out base
          modification calls from the start and end of the reads. For example, 4,8 will filter out
          base modification calls in the first 4 and last 8 bases of the read.

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification calls at the ends of
          reads, only _keep_ base modification calls at the ends of reads. E.g. if usually, &quot;4,8&quot;
          would remove (i.e. filter out) base modification calls in the first 4 and last 8 bases of
          the read, using this flag will keep only base modification calls in the first 4 and last 8
          bases.

      --hist
          Output histogram of base modification prediction probabilities.

      --buckets &lt;BUCKETS&gt;
          Number of buckets for the histogram, if used.
          
          [default: 128]

  -n, --num-reads &lt;NUM_READS&gt;
          Approximate maximum number of reads to use, especially recommended when using a large BAM
          without an index. If an indexed BAM is provided, the reads will be sampled evenly over the
          length of the aligned reference. If a region is passed with the --region option, they will
          be sampled over the genomic region. Actual number of reads used may deviate slightly from
          this number.
          
          [default: 10042]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Instead of using a defined number of reads, specify a fraction of reads to sample, for
          example 0.1 will sample 1/10th of the reads.

      --no-sampling
          No sampling, use all of the reads to calculate the filter thresholds.

  -s, --seed &lt;SEED&gt;
          Random seed for deterministic running, the default is non-deterministic, only used when no
          BAM index is provided.

      --region &lt;REGION&gt;
          Process only the specified region of the BAM when collecting probabilities. Format should
          be &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;.

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently. Smaller interval chunk sizes
          will use less memory but incur more overhead. Only used when sampling probs from an
          indexed bam.
          
          [default: 1000000]

      --include-bed &lt;INCLUDE_BED&gt;
          Only sample base modification probabilities that are aligned to the positions in this BED
          file. (alias: include-positions)

      --only-mapped
          Only use base modification probabilities that are aligned (i.e. ignore soft-clipped, and
          inserted bases).

  -h, --help
          Print help information (use `-h` for a summary).
</code></pre>
<h2 id="summary"><a class="header" href="#summary">summary</a></h2>
<pre><code class="language-text">Summarize the mod tags present in a BAM and get basic statistics. The default output is a totals
table (designated by '#' lines) and a modification calls table. Descriptions of the columns can be
found in the README.

Usage: modkit summary [OPTIONS] &lt;IN_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input modBam, can be a path to a file or one of `-` or `stdin` to specify a stream from
          standard input.

Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use.
          
          [default: 4]

      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them. Setting a file is
          recommended.

      --tsv
          Output summary as a tab-separated variables stdout instead of a table.

      --suppress-progress
          Hide the progress bar.

  -n, --num-reads &lt;NUM_READS&gt;
          Approximate maximum number of reads to use, especially recommended when using a large BAM
          without an index. If an indexed BAM is provided, the reads will be sampled evenly over the
          length of the aligned reference. If a region is passed with the --region option, they will
          be sampled over the genomic region. Actual number of reads used may deviate slightly from
          this number.
          
          [default: 10042]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Instead of using a defined number of reads, specify a fraction of reads to sample when
          estimating the filter threshold. For example 0.1 will sample 1/10th of the reads.

      --no-sampling
          No sampling, use all of the reads to calculate the filter thresholds and generating the
          summary.

  -s, --seed &lt;SEED&gt;
          Sets a random seed for deterministic running (when using --sample-frac), the default is
          non-deterministic, only used when no BAM index is provided.

      --no-filtering
          Do not perform any filtering, include all base modification calls in the summary. See
          filtering.md for details on filtering.

  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted variant is below
          this confidence percentile. For example, 0.1 will filter out the 10% lowest confidence
          base modification calls.
          
          [default: 0.1]

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or per-base. Global filter threshold can be
          specified with by a decimal number (e.g. 0.75). Per-base thresholds can be specified by
          colon-separated values, for example C:0.75 specifies a threshold value of 0.75 for
          cytosine modification calls. Additional per-base thresholds can be specified by repeating
          the option: for example --filter-threshold C:0.75 --filter-threshold A:0.70 or specify a
          single base option and a default for all other bases with: --filter-threshold A:0.70
          --filter-threshold 0.9 will specify a threshold value of 0.70 for adenine and 0.9 for all
          other base modification calls.

      --mod-thresholds &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification, independent of the threshold
          for the primary sequence base or the default. For example, to set the pass threshold for
          5hmC to 0.8 use `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless the
          `--filter-threshold` option is also passed. See the online documentation for more details.

      --ignore &lt;IGNORE&gt;
          Ignore a modified base class  _in_situ_ by redistributing base modification probability
          equally across other options. For example, if collapsing 'h', with 'm' and canonical
          options, half of the probability of 'h' will be added to both 'm' and 'C'. A full
          description of the methods can be found in collapse.md.

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the start or the end of the
          read. Two comma-separated values may be provided to asymmetrically filter out base
          modification calls from the start and end of the reads. For example, 4,8 will filter out
          base modification calls in the first 4 and last 8 bases of the read.

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification calls at the ends of
          reads, only _keep_ base modification calls at the ends of reads. E.g. if usually, &quot;4,8&quot;
          would remove (i.e. filter out) base modification calls in the first 4 and last 8 bases of
          the read, using this flag will keep only base modification calls in the first 4 and last 8
          bases.

      --include-bed &lt;INCLUDE_BED&gt;
          Only summarize base modification probabilities that are aligned to the positions in this
          BED file. (alias: include-positions)

      --only-mapped
          Only use base modification probabilities that are aligned (i.e. ignore soft-clipped, and
          inserted bases).

      --region &lt;REGION&gt;
          Process only the specified region of the BAM when collecting probabilities. Format should
          be &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;.

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          When using regions, interval chunk size in base pairs to process concurrently. Smaller
          interval chunk sizes will use less memory but incur more overhead.
          
          [default: 1000000]

  -h, --help
          Print help information (use `-h` for a summary).
</code></pre>
<h2 id="motif-bed"><a class="header" href="#motif-bed">motif-bed</a></h2>
<pre><code class="language-text">Create BED file with all locations of a sequence motif. Example: modkit motif-bed CG 0

Usage: modkit motif-bed [OPTIONS] &lt;FASTA&gt; &lt;MOTIF&gt; &lt;OFFSET&gt;

Arguments:
  &lt;FASTA&gt;   Input FASTA file.
  &lt;MOTIF&gt;   Motif to search for within FASTA, e.g. CG.
  &lt;OFFSET&gt;  Offset within motif, e.g. 0.

Options:
  -k, --mask  Respect soft masking in the reference FASTA.
  -h, --help  Print help information.
</code></pre>
<h2 id="call-mods"><a class="header" href="#call-mods">call-mods</a></h2>
<pre><code class="language-text">Call mods from a modBam, creates a new modBam with probabilities set to 100% if a base modification
is called or 0% if called canonical.

Usage: modkit call-mods [OPTIONS] &lt;IN_BAM&gt; &lt;OUT_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input BAM, may be sorted and have associated index available. Can be a path to a file or
          one of `-` or `stdin` to specify a stream from standard input.

  &lt;OUT_BAM&gt;
          Output BAM, can be a path to a file or one of `-` or `stdin` to specify a stream from
          standard input.

Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them. Setting a file is
          recommended.

      --ff
          Fast fail, stop processing at the first invalid sequence record. Default behavior is to
          continue and report failed/skipped records at the end.

      --suppress-progress
          Hide the progress bar.

  -t, --threads &lt;THREADS&gt;
          Number of threads to use while processing chunks concurrently.
          
          [default: 4]

  -n, --num-reads &lt;NUM_READS&gt;
          Sample approximately this many reads when estimating the filtering threshold. If
          alignments are present reads will be sampled evenly across aligned genome. If a region is
          specified, either with the --region option or the --sample-region option, then reads will
          be sampled evenly across the region given. This option is useful for large BAM files. In
          practice, 10-50 thousand reads is sufficient to estimate the model output distribution and
          determine the filtering threshold.
          
          [default: 10042]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Sample this fraction of the reads when estimating the filter-percentile. In practice,
          50-100 thousand reads is sufficient to estimate the model output distribution and
          determine the filtering threshold. See filtering.md for details on filtering.

      --seed &lt;SEED&gt;
          Set a random seed for deterministic running, the default is non-deterministic, only used
          when no BAM index is provided.

      --sample-region &lt;SAMPLE_REGION&gt;
          Specify a region for sampling reads from when estimating the threshold probability. If
          this option is not provided, but --region is provided, the genomic interval passed to
          --region will be used. Format should be &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;.

      --sampling-interval-size &lt;SAMPLING_INTERVAL_SIZE&gt;
          Interval chunk size to process concurrently when estimating the threshold probability, can
          be larger than the pileup processing interval.
          
          [default: 1000000]

  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted variant is below
          this confidence percentile. For example, 0.1 will filter out the 10% lowest confidence
          modification calls.
          
          [default: 0.1]

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or per primary base. A global filter threshold can
          be specified with by a decimal number (e.g. 0.75). Per-base thresholds can be specified by
          colon-separated values, for example C:0.75 specifies a threshold value of 0.75 for
          cytosine modification calls. Additional per-base thresholds can be specified by repeating
          the option: for example --filter-threshold C:0.75 --filter-threshold A:0.70 or specify a
          single base option and a default for all other bases with: --filter-threshold A:0.70
          --filter-threshold 0.9 will specify a threshold value of 0.70 for adenine and 0.9 for all
          other base modification calls.

      --mod-threshold &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification, independent of the threshold
          for the primary sequence base or the default. For example, to set the pass threshold for
          5hmC to 0.8 use `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless the
          `--filter-threshold` option is also passed. See the online documentation for more details.

      --no-filtering
          Don't filter base modification calls, assign each base modification to the highest
          probability prediction.

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the start or the end of the
          read. Two comma-separated values may be provided to asymmetrically filter out base
          modification calls from the start and end of the reads. For example, 4,8 will filter out
          base modification calls in the first 4 and last 8 bases of the read.

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification calls at the ends of
          reads, only _keep_ base modification calls at the ends of reads. E.g. if usually, &quot;4,8&quot;
          would remove (i.e. filter out) base modification calls in the first 4 and last 8 bases of
          the read, using this flag will keep only base modification calls in the first 4 and last 8
          bases.

      --output-sam
          Output SAM format instead of BAM.

  -h, --help
          Print help information (use `-h` for a summary).
</code></pre>
<h2 id="extract"><a class="header" href="#extract">extract</a></h2>
<pre><code class="language-text">Extract read-level base modification information from a modBAM into a tab-separated values table.

Usage: modkit extract [OPTIONS] &lt;IN_BAM&gt; &lt;OUT_PATH&gt;

Arguments:
  &lt;IN_BAM&gt;
          Path to modBAM file to extract read-level information from, or one of `-` or `stdin` to
          specify a stream from standard input. If a file is used it may be sorted and have
          associated index.

  &lt;OUT_PATH&gt;
          Path to output file, &quot;stdout&quot; or &quot;-&quot; will direct output to standard out. Specifying &quot;null&quot;
          will not output the extract table (useful if all you need is the `--read-calls` output
          table).

Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use.
          
          [default: 4]

      --log-filepath &lt;LOG_FILEPATH&gt;
          Path to file to write run log, setting this file is recommended.

      --mapped-only
          Include only mapped bases in output. (alias: mapped)

      --allow-non-primary
          Output aligned secondary and supplementary base modification probabilities as additional
          rows. The primary alignment will have all of the base modification probabilities
          (including soft-clipped ones, unless --mapped-only is used). The non-primary alignments
          will only have mapped bases in the output.

      --num-reads &lt;NUM_READS&gt;
          Number of reads to use. Note that when using a sorted, indexed modBAM that the sampling
          algorithm will attempt to sample records evenly over the length of the reference sequence.
          The result is the final number of records used may be slightly more or less than the
          requested number. When piping from stdin or using a modBAM without an index, the requested
          number of reads will be exact.

      --region &lt;REGION&gt;
          Process only reads that are aligned to a specified region of the BAM. Format should be
          &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;.

      --force
          Force overwrite of output file.

      --suppress-progress
          Hide the progress bar.

      --kmer-size &lt;KMER_SIZE&gt;
          Set the query and reference k-mer size (if a reference is provided). Maxumum number for
          this value is 12.
          
          [default: 5]

      --ignore-index
          Ignore the BAM index (if it exists) and default to a serial scan of the BAM.

      --read-calls-path &lt;READ_CALLS_PATH&gt;
          Produce a table of read-level base modification calls. This table has, for each read, one
          row for each base modification call in that read using the same thresholding algorithm as
          in pileup, or summary (see online documentation for details on thresholds). Passing this
          option will cause `modkit` to estimate the pass thresholds from the data unless a
          `--filter-threshold` value is passed to the command. (alias: --read-calls)

      --reference &lt;REFERENCE&gt;
          Path to reference FASTA to extract reference context information from. If no reference is
          provided, `ref_kmer` column will be &quot;.&quot; in the output. (alias: ref)

      --include-bed &lt;INCLUDE_BED&gt;
          BED file with regions to include (alias: include-positions). Implicitly only includes
          mapped sites

  -v, --exclude-bed &lt;EXCLUDE_BED&gt;
          BED file with regions to _exclude_ (alias: exclude).

      --motif &lt;MOTIF&gt; &lt;MOTIF&gt;
          Output read-level base modification probabilities restricted to the reference sequence
          motifs provided. The first argument should be the sequence motif and the second argument
          is the 0-based offset to the base to pileup base modification counts for. For example:
          --motif CGCG 0 indicates include base modifications for which the read is aligned to the
          first C on the top strand and the last C (complement to G) on the bottom strand. The --cpg
          argument is short hand for --motif CG 0. This argument can be passed multiple times.

      --cpg
          Only output counts at CpG motifs. Requires a reference sequence to be provided.

  -k, --mask
          When using motifs, respect soft masking in the reference sequence.

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or per-base. Global filter threshold can be
          specified with by a decimal number (e.g. 0.75). Per-base thresholds can be specified by
          colon-separated values, for example C:0.75 specifies a threshold value of 0.75 for
          cytosine modification calls. Additional per-base thresholds can be specified by repeating
          the option: for example --filter-threshold C:0.75 --filter-threshold A:0.70 or specify a
          single base option and a default for all other bases with: --filter-threshold A:0.70
          --filter-threshold 0.9 will specify a threshold value of 0.70 for adenine and 0.9 for all
          other base modification calls.

      --mod-thresholds &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification, independent of the threshold
          for the primary sequence base or the default. For example, to set the pass threshold for
          5hmC to 0.8 use `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless the
          `--filter-threshold` option is also passed. See the online documentation for more details.

      --no-filtering
          Do not perform any filtering, include all mod base calls in output. See filtering.md for
          details on filtering.

      --sampling-interval-size &lt;SAMPLING_INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently when estimating the threshold
          probability.
          
          [default: 1000000]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Sample this fraction of the reads when estimating the pass-threshold. In practice, 10-100
          thousand reads is sufficient to estimate the model output distribution and determine the
          filtering threshold. See filtering.md for details on filtering.

  -n, --sample-num-reads &lt;SAMPLE_NUM_READS&gt;
          Sample this many reads when estimating the filtering threshold. If a sorted, indexed
          modBAM is provided reads will be sampled evenly across aligned genome. If a region is
          specified, with the --region, then reads will be sampled evenly across the region given.
          This option is useful for large BAM files. In practice, 10-50 thousand reads is sufficient
          to estimate the model output distribution and determine the filtering threshold.
          
          [default: 10042]

      --seed &lt;SEED&gt;
          Set a random seed for deterministic running, the default is non-deterministic.

  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted variant is below
          this confidence percentile. For example, 0.1 will filter out the 10% lowest confidence
          modification calls.
          
          [default: 0.1]

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the start or the end of the
          read. Two comma-separated values may be provided to asymmetrically filter out base
          modification calls from the start and end of the reads. For example, 4,8 will filter out
          base modification calls in the first 4 and last 8 bases of the read.

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification calls at the ends of
          reads, only _keep_ base modification calls at the ends of reads. E.g. if usually, &quot;4,8&quot;
          would remove (i.e. filter out) base modification calls in the first 4 and last 8 bases of
          the read, using this flag will keep only base modification calls in the first 4 and last 8
          bases.

      --ignore &lt;IGNORE&gt;
          Ignore a modified base class  _in_situ_ by redistributing base modification probability
          equally across other options. For example, if collapsing 'h', with 'm' and canonical
          options, half of the probability of 'h' will be added to both 'm' and 'C'. A full
          description of the methods can be found in collapse.md.

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently. Smaller interval chunk sizes
          will use less memory but incur more overhead. Only used when an indexed modBam is provided.
          
          [default: 100000]

      --ignore-implicit
          Ignore implicitly canonical base modification calls. When the `.` flag is used in the MM
          tag, this implies that bases missing a base modification probability are to be assumed
          canonical. Set this flag to omit those base modifications from the output. For additional
          details see the SAM spec: https://samtools.github.io/hts-specs/SAMtags.pdf.

  -h, --help
          Print help information (use `-h` for a summary).
</code></pre>
<h2 id="repair"><a class="header" href="#repair">repair</a></h2>
<pre><code class="language-text">Repair MM and ML tags in one bam with the correct tags from another. To use this command, both
modBAMs _must_ be sorted by read name. The &quot;donor&quot; modBAM's reads must be a superset of the
acceptor's reads. Extra reads in the donor are allowed, and multiple reads with the same name
(secondary, etc.) are allowed in the acceptor. Reads with an empty SEQ field cannot be repaired and
will be rejected. Reads where there is an ambiguous alignment of the acceptor to the donor will be
rejected (and logged). See the full documentation for details.

Usage: modkit repair [OPTIONS] --donor-bam &lt;DONOR_BAM&gt; --acceptor-bam &lt;ACCEPTOR_BAM&gt; --output-bam &lt;OUTPUT_BAM&gt;

Options:
  -d, --donor-bam &lt;DONOR_BAM&gt;        Donor modBAM with original MM/ML tags. Must be sorted by read
                                     name.
  -a, --acceptor-bam &lt;ACCEPTOR_BAM&gt;  Acceptor modBAM with reads to have MM/ML base modification data
                                     projected on to. Must be sorted by read name.
  -o, --output-bam &lt;OUTPUT_BAM&gt;      output modBAM location.
      --log-filepath &lt;LOG_FILEPATH&gt;  File to write logs to, it is recommended to use this option as
                                     some reads may be rejected and logged here
  -t, --threads &lt;THREADS&gt;            The number of threads to use [default: 4]
  -h, --help                         Print help information.
</code></pre>
<h2 id="pileup-hemi"><a class="header" href="#pileup-hemi">pileup-hemi</a></h2>
<pre><code class="language-text">Tabulates double-stranded base modification patters (such as hemi-methylation) across genomic motif
positions. This command produces a bedMethyl file, the schema can be found in the online
documentation.

Usage: modkit pileup-hemi [OPTIONS] &lt;IN_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input BAM, should be sorted and have associated index available.

Options:
  -o, --out-bed &lt;OUT_BED&gt;
          Output file to write results into. Will write to stdout if not provided.

      --cpg
          Aggregate double-stranded base modifications for CpG dinucleotides. This flag is
          short-hand for --motif CG 0.

      --motif &lt;MOTIF&gt; &lt;MOTIF&gt;
          Specify the sequence motif to pileup double-stranded base modification pattern counts for.
          The first argument should be the sequence motif and the second argument is the 0-based
          offset to the base to pileup base modification counts for. For example: --motif CG 0
          indicates to generate pattern counts for the C on the top strand and the following C
          (opposite to G) on the negative strand. The motif must be reverse-complement palindromic
          or an error will be raised. See the documentation for more examples and details.

  -r, --ref &lt;REFERENCE_FASTA&gt;
          Reference sequence in FASTA format.

      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them. Setting a file is
          recommended. (alias: log)

      --region &lt;REGION&gt;
          Process only the specified region of the BAM when performing pileup. Format should be
          &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;. Commas are allowed.

      --max-depth &lt;MAX_DEPTH&gt;
          Maximum number of records to use when calculating pileup. This argument is passed to the
          pileup engine. If you have high depth data, consider increasing this value substantially.
          Must be less than 2147483647 or an error will be raised.
          
          [default: 8000]

  -t, --threads &lt;THREADS&gt;
          Number of threads to use while processing chunks concurrently.
          
          [default: 4]

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently. Smaller interval chunk sizes
          will use less memory but incur more overhead.
          
          [default: 100000]

      --chunk-size &lt;CHUNK_SIZE&gt;
          Break contigs into chunks containing this many intervals (see `interval_size`). This
          option can be used to help prevent excessive memory usage, usually with no performance
          penalty. By default, modkit will set this value to 1.5x the number of threads specified,
          so if 4 threads are specified the chunk_size will be 6. A warning will be shown if this
          option is less than the number of threads specified.

      --suppress-progress
          Hide the progress bar.

  -n, --num-reads &lt;NUM_READS&gt;
          Sample this many reads when estimating the filtering threshold. Reads will be sampled
          evenly across aligned genome. If a region is specified, either with the --region option or
          the --sample-region option, then reads will be sampled evenly across the region given.
          This option is useful for large BAM files. In practice, 10-50 thousand reads is sufficient
          to estimate the model output distribution and determine the filtering threshold.
          
          [default: 10042]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Sample this fraction of the reads when estimating the filter-percentile. In practice,
          50-100 thousand reads is sufficient to estimate the model output distribution and
          determine the filtering threshold. See filtering.md for details on filtering.

      --seed &lt;SEED&gt;
          Set a random seed for deterministic running, the default is non-deterministic.

      --no-filtering
          Do not perform any filtering, include all mod base calls in output. See filtering.md for
          details on filtering.

  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted variant is below
          this confidence percentile. For example, 0.1 will filter out the 10% lowest confidence
          modification calls.
          
          [default: 0.1]

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or per-base. Global filter threshold can be
          specified with by a decimal number (e.g. 0.75). Per-base thresholds can be specified by
          colon-separated values, for example C:0.75 specifies a threshold value of 0.75 for
          cytosine modification calls. Additional per-base thresholds can be specified by repeating
          the option: for example --filter-threshold C:0.75 --filter-threshold A:0.70 or specify a
          single base option and a default for all other bases with: --filter-threshold A:0.70
          --filter-threshold 0.9 will specify a threshold value of 0.70 for adenine and 0.9 for all
          other base modification calls.

      --mod-thresholds &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification, independent of the threshold
          for the primary sequence base or the default. For example, to set the pass threshold for
          5hmC to 0.8 use `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless the
          `--filter-threshold` option is also passed. See the online documentation for more details.

      --sample-region &lt;SAMPLE_REGION&gt;
          Specify a region for sampling reads from when estimating the threshold probability. If
          this option is not provided, but --region is provided, the genomic interval passed to
          --region will be used. Format should be &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;.

      --sampling-interval-size &lt;SAMPLING_INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently when estimating the threshold
          probability, can be larger than the pileup processing interval.
          
          [default: 1000000]

      --include-bed &lt;INCLUDE_BED&gt;
          BED file that will restrict threshold estimation and pileup results to positions
          overlapping intervals in the file. (alias: include-positions)

      --include-unmapped
          Include unmapped base modifications when estimating the pass threshold.

      --ignore &lt;IGNORE&gt;
          Ignore a modified base class  _in_situ_ by redistributing base modification probability
          equally across other options. For example, if collapsing 'h', with 'm' and canonical
          options, half of the probability of 'h' will be added to both 'm' and 'C'. A full
          description of the methods can be found in collapse.md.

      --force-allow-implicit
          Force allow implicit-canonical mode. By default modkit does not allow pileup with the
          implicit mode (e.g. C+m, no '.' or '?'). The `update-tags` subcommand is provided to
          update tags to the new mode. This option allows the interpretation of implicit mode tags:
          residues without modified base probability will be interpreted as being the non-modified
          base.

  -k, --mask
          Respect soft masking in the reference FASTA.

      --combine-mods
          Combine base modification calls, all counts of modified bases are summed together. See
          collapse.md for details.

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the start or the end of the
          read. Two comma-separated values may be provided to asymmetrically filter out base
          modification calls from the start and end of the reads. For example, 4,8 will filter out
          base modification calls in the first 4 and last 8 bases of the read.

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification calls at the ends of
          reads, only _keep_ base modification calls at the ends of reads. E.g. if usually, &quot;4,8&quot;
          would remove (i.e. filter out) base modification calls in the first 4 and last 8 bases of
          the read, using this flag will keep only base modification calls in the first 4 and last 8
          bases.

      --only-tabs
          Separate bedMethyl columns with only tabs. The default is to use tabs for the first 10
          fields and spaces thereafter. The default behavior is more likely to be compatible with
          genome viewers. Enabling this option may make it easier to parse the output with tabular
          data handlers that expect a single kind of separator.

  -h, --help
          Print help information (use `-h` for a summary).
</code></pre>
<h2 id="pileup-hemi-pair"><a class="header" href="#pileup-hemi-pair">pileup-hemi <code>pair</code></a></h2>
<pre><code class="language-text">Compare regions in a pair of samples (for example, tumor and normal or control and experiment). A
sample is input as a bgzip pileup bedMethyl (produced by pileup, for example) that has an associated
tabix index. Output is a BED file with the score column indicating the magnitude of the difference
in methylation between the two samples. See the online documentation for additional details.

Usage: modkit dmr pair [OPTIONS] -a &lt;CONTROL_BED_METHYL&gt; -b &lt;EXP_BED_METHYL&gt; --ref &lt;REFERENCE_FASTA&gt;

Options:
  -a &lt;CONTROL_BED_METHYL&gt;
          Bgzipped bedMethyl file for the first (usually control) sample. There should be a tabix
          index with the same name and .tbi next to this file or the --index-a option must be
          provided.
  -b &lt;EXP_BED_METHYL&gt;
          Bgzipped bedMethyl file for the second (usually experimental) sample. There should be a
          tabix index with the same name and .tbi next to this file or the --index-b option must be
          provided.
  -o, --out-path &lt;OUT_PATH&gt;
          Path to file to direct output, optional, no argument will direct output to stdout.
  -r, --regions-bed &lt;REGIONS_BED&gt;
          BED file of regions over which to compare methylation levels. Should be tab-separated
          (spaces allowed in the &quot;name&quot; column). Requires chrom, chromStart and chromEnd. The Name
          column is optional. Strand is currently ignored. When omitted, methylation levels are
          compared at each site in the `-a`/`control_bed_methyl` BED file (or optionally, the
          `-b`/`exp_bed_methyl` file with the `--use-b` flag.
      --use-b
          When performing site-level DMR, use the bedMethyl indicated by the -b/exp_bed_methyl
          argument to collect bases to score.
      --ref &lt;REFERENCE_FASTA&gt;
          Path to reference fasta for the pileup.
  -m &lt;MODIFIED_BASES&gt;
          Bases to use to calculate DMR, may be multiple. For example, to calculate differentially
          methylated regions using only cytosine modifications use --base C.
      --log-filepath &lt;LOG_FILEPATH&gt;
          File to write logs to, it's recommended to use this option.
  -t, --threads &lt;THREADS&gt;
          Number of threads to use [default: 4]
      --batch-size &lt;BATCH_SIZE&gt;
          Control the  batch size. The batch size is the number of regions to load at a time. Each
          region will be processed concurrently. Loading more regions at a time will decrease IO to
          load data, but will use more memory. Default will be 50% more than the number of threads
          assigned.
  -k, --mask
          Respect soft masking in the reference FASTA.
      --suppress-progress
          Don't show progress bars.
  -f, --force
          Force overwrite of output file, if it already exists.
      --index-a &lt;INDEX_A&gt;
          Path to tabix index associated with -a (--control-bed-methyl) bedMethyl file.
      --index-b &lt;INDEX_B&gt;
          Path to tabix index associated with -b (--exp-bed-methyl) bedMethyl file.
      --missing &lt;HANDLE_MISSING&gt;
          How to handle regions found in the `--regions` BED file. quiet =&gt; ignore regions that are
          not found in the tabix header warn =&gt; log (debug) regions that are missing fatal =&gt; log
          (error) and exit the program when a region is missing. [default: warn] [possible values:
          quiet, warn, fail]
      --min-valid-coverage &lt;MIN_VALID_COVERAGE&gt;
          Minimum valid coverage required to use an entry from a bedMethyl. See the help for pileup
          for the specification and description of valid coverage. [default: 0]
  -h, --help
          Print help information.
</code></pre>
<h2 id="pileup-hemi-1"><a class="header" href="#pileup-hemi-1">pileup-hemi</a></h2>
<pre><code class="language-text">Compare regions between all pairs of samples (for example a trio sample set or haplotyped trio
sample set). As with `pair` all inputs must be bgzip compressed bedMethyl files with associated
tabix indices. Each sample must be assigned a name. Output is a directory of BED files with the
score column indicating the magnitude of the difference in methylation between the two samples
indicated in the file name. See the online documentation for additional details.

Usage: modkit dmr multi [OPTIONS] --out-dir &lt;OUT_DIR&gt; --ref &lt;REFERENCE_FASTA&gt;

Options:
  -s, --sample &lt;SAMPLES&gt; &lt;SAMPLES&gt;
          Two or more named samples to compare. Two arguments are required &lt;path&gt; &lt;name&gt;. This
          option should be repeated at least two times.
  -i, --index &lt;INDICES&gt; &lt;INDICES&gt;
          Optional, paths to tabix indices associated with named samples. Two arguments are required
          &lt;path&gt; &lt;name&gt; where &lt;name&gt; corresponds to the name of the sample given to the -s/--sample
          argument.
  -r, --regions-bed &lt;REGIONS_BED&gt;
          BED file of regions over which to compare methylation levels. Should be tab-separated
          (spaces allowed in the &quot;name&quot; column). Requires chrom, chromStart and chromEnd. The Name
          column is optional. Strand is currently ignored. When omitted, methylation levels are
          compared at each site in common between the two bedMethyl files being compared.
  -o, --out-dir &lt;OUT_DIR&gt;
          Directory to place output DMR results in BED format.
  -p, --prefix &lt;PREFIX&gt;
          Prefix files in directory with this label.
      --ref &lt;REFERENCE_FASTA&gt;
          Path to reference fasta for the pileup.
  -m &lt;MODIFIED_BASES&gt;
          Bases to use to calculate DMR, may be multiple. For example, to calculate differentially
          methylated regions using only cytosine modifications use --base C.
      --log-filepath &lt;LOG_FILEPATH&gt;
          File to write logs to, it's recommended to use this option.
  -t, --threads &lt;THREADS&gt;
          Number of threads to use. [default: 4]
  -k, --mask
          Respect soft masking in the reference FASTA.
      --suppress-progress
          Don't show progress bars.
  -f, --force
          Force overwrite of output file, if it already exists
      --missing &lt;HANDLE_MISSING&gt;
          How to handle regions found in the `--regions` BED file. quiet =&gt; ignore regions that are
          not found in the tabix header warn =&gt; log (debug) regions that are missing fatal =&gt; log
          (error) and exit the program when a region is missing. [default: warn] [possible values:
          quiet, warn, fail]
      --min-valid-coverage &lt;MIN_VALID_COVERAGE&gt;
          Minimum valid coverage required to use an entry from a bedMethyl. See the help for pileup
          for the specification and description of valid coverage. [default: 0]
  -h, --help
          Print help information
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="troubleshooting"><a class="header" href="#troubleshooting">Troubleshooting</a></h1>
<p>It's recommended to run all <code>modkit</code> commands with the <code>--log-filepath &lt;path-to-file&gt;</code>
option set. When unexpected outputs are produced inspecting this file will often indicate
the reason.</p>
<h2 id="missing-secondary-and-supplementary-alignments-in-output"><a class="header" href="#missing-secondary-and-supplementary-alignments-in-output">Missing secondary and supplementary alignments in output</a></h2>
<p>As of v0.2.4 secondary and supplementary alignments are supported in <code>adjust-mods</code>, <code>update-tags</code>, <code>call-mods</code>, and (optionally) in <code>extract</code>.
However, in order to use these alignment records correctly, the <code>MN</code> tag must be present and correct in the record. 
The <code>MN</code> tag indicates the length of the sequence corresponding to the <code>MM</code> and <code>ML</code> tags.
As of dorado v0.5.0 the <code>MN</code> tag is output when modified base calls are produced.
If the aligner has hard-clipped the sequence, this number will not match the sequence length and the record cannot be used. 
Similarly, if the SEQ field is empty (sequence length zero), the record cannot be used. 
One way to use supplementary alignments is to specify the <code>-Y</code> flag when using <a href="https://github.com/nanoporetech/dorado/">dorado</a> or <a href="https://lh3.github.io/minimap2/minimap2.html">minimap2</a>. 
For these programs, when <code>-Y</code> is specified, the sequence will not be hardclipped in supplementary alignments and will be present in secondary alignments. 
Other mapping algorithms that are &quot;MM tag-aware&quot; may allow hard-clipping and update the <code>MM</code> and <code>ML</code> tags, <code>modkit</code> will accept these records as long as the <code>MN</code> tag indicates the correct sequence length.</p>
<h2 id="no-rows-in-modkit-pileup-output"><a class="header" href="#no-rows-in-modkit-pileup-output">No rows in <code>modkit pileup</code> output.</a></h2>
<p>First, check the logfile, there may be many lines with a variant of</p>
<pre><code class="language-text">record 905b4cb4-15e2-4060-9c98-866d0aff78bb has un-allowed mode
(ImplicitProbModified), use '--force-allow-implicit' or 'modkit update-tags --mode
ambiguous
</code></pre>
<p>As suggested, using <a href="./intro_adjust.html##updating-the-flag--and-">update</a> or setting the
<code>--force-allow-implicit</code> flag should produce output from these records.</p>
<p>If the MM tag contains an un-supported mod-code (a <a href="./limitations.html">limitation</a> that
will be removed in a future release), these errors will be logged. To process the modBAM,
first <a href="./intro_adjust.html#changing-the-base-modification-code">convert</a> the tags.</p>
<p>In general, if there are MM/ML tags in the modBAM and the bedMethyl is still empty the
log file will contain a line explaining why each record is being skipped.</p>
<h2 id="not-sampling-enough-reads-to-estimate-threshold"><a class="header" href="#not-sampling-enough-reads-to-estimate-threshold">Not sampling enough reads to estimate threshold.</a></h2>
<p>If you have previously downsampled the modBAM to a specific region, for example with a
command like</p>
<pre><code>samtools view -bh input.sorted.bam chr1
</code></pre>
<p>modkit will still try and sample reads evenly across the entire genome but fail to find
any aligned to other contigs. To remedy this, either pass the same <code>--region chr1</code> option
or set the <code>--sampling-frac 1.0</code>. The former will set <code>modkit</code> to only use the region
present in the BAM. For example,</p>
<pre><code>modkit pileup input.bam output.bed --region chr1
# or
modkit pileup input.bam output.bed --sample-region chr1
</code></pre>
<p>The latter will sample all of the reads in the BAM, which may slow
down the process, so in general the best advice is to either use the same <code>--region &lt;contig&gt;</code> when using <code>modkit</code> or don't downsample the BAM ahead of time (and use <code>--region &lt;contig&gt;</code> in order to get the desired bedMethyl).</p>
<p>The <code>summary</code>, <code>sample-probs</code>, and <code>pileup</code> subcommands all use the same <code>--region</code>
option.</p>
<h2 id="cg-positions-are-missing-from-output-bedmethyl"><a class="header" href="#cg-positions-are-missing-from-output-bedmethyl">CG positions are missing from output bedMethyl.</a></h2>
<p>When running with <code>--preset traditional</code> or <code>--cpg</code> the resultant bedMethyl file will only
contains CG positions. However, it will not include positions for which the pass coverage
is zero (see <a href="./intro_bedmethyl.html#description-of-bedmethyl-output">the column
descriptions</a>). This is to be
expected.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="current-limitations"><a class="header" href="#current-limitations">Current limitations</a></h1>
<p>Known limitations and forecasts for when they will be removed.</p>
<ol>
<li>Ambiguous DNA bases in ML tags are not supported (for example <code>N+m?</code>).
<ul>
<li>This limitation will be removed in version 0.2.z</li>
</ul>
</li>
<li>During <code>modkit pileup</code>, it is assumed that each read should only have one primary alignment. If a read name
is detected more than once, the occurrence is logged but both alignments will be used. This limitation may be
removed in the future with a form of dynamic de-duplication.</li>
<li>Only one MM-flag (<code>.</code>, <code>?</code>) per-canonical base is supported within a read.
<ul>
<li>This limitation may be removed in the future.</li>
</ul>
</li>
</ol>
<div style="break-before: page; page-break-before: always;"></div><h1 id="performance-considerations-1"><a class="header" href="#performance-considerations-1">Performance considerations</a></h1>
<h2 id="sharding-a-large-modbam-by-region"><a class="header" href="#sharding-a-large-modbam-by-region">Sharding a large modBAM by region.</a></h2>
<p>The <code>--region</code> option in <code>pileup</code>, <code>summary</code>, and <code>sample-probs</code> can be used to
operate on a subset of records in a large BAM. If you're working in a
distributed environment, the genome could be sharded into large sections which
are specified to <code>modkit</code> in concurrent processes and merged afterward in a
&quot;map-reduce&quot; pattern.</p>
<h2 id="setting-the---interval-size-and---chunk-size-pileup"><a class="header" href="#setting-the---interval-size-and---chunk-size-pileup">Setting the <code>--interval-size</code> and <code>--chunk-size</code> (<code>pileup</code>).</a></h2>
<p>Whenever operating on a sorted, indexed BAM, <code>modkit</code> will operate in parallel
on disjoint spans of the genome. The length of these spans (i.e. intervals) can
be determined by the <code>--interval-size</code>  or the <code>--sampling-interval-size</code> (for
the sampling algorithm only). The defaults for these parameters works well for
genomes such as the human genome. For smaller genomes with high coverage, you
may decide to <em>decrease</em> the interval size in order to take advantage of
parallelism. The <code>pileup</code> subcommand also has a <code>--chunk-size</code> option that will
limit the total number of <em>intervals</em> computed on in parallel. By default,
<code>modkit</code> will set this parameter to be 50% larger than the number of threads.
In general, this is a good setting for balancing parallelism and memory usage.
Increasing the <code>--chunk-size</code> can increase parallelism (and decrease run time)
but will consume more memory.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="algorithm-details"><a class="header" href="#algorithm-details">Algorithm details</a></h1>
<ul>
<li><a href="./filtering.html">Filtering low confidence base modification calls</a></li>
<li><a href="./filtering.html">Removing and combining modified base types</a></li>
</ul>
<div style="break-before: page; page-break-before: always;"></div><h1 id="partitioning-pass-and-fail-base-modification-calls"><a class="header" href="#partitioning-pass-and-fail-base-modification-calls">Partitioning pass and fail base modification calls.</a></h1>
<p>Base modification calls can be removed if they are low confidence based on the predicted modification probabilities. 
In general, <code>modkit</code> will estimate a pass confidence threshold value based on the input data. 
Threshold values for modifications on a primary sequence base can be specified on the command line with the <code>--filter-threshold</code> option. 
For example to set a threshold for cytosine modifications at 0.8 and adenine modifications at 0.9 provide <code>--filter-threshold C:0.8 --filter-threshold A:0.9</code>.
Pass threshold values per base modification can also be specified.
For example, to specify a threshold for canonical adenine at 0.8 and 6mA at 0.9 use <code>--filter-threshold A:0.8 --mod-thresholds a:0.9</code>.
Or to specify a threshold of 0.8 for 5mC, 0.9 for 5hmC, and 0.85 for canonical cytosine: <code>--filter-threshold C:0.85 --mod-thresholds m:0.8 --mod-thresholds h:0.9</code></p>
<p>Keep in mind that the <code>--mod-threshold</code> option will treat <code>A</code>, <code>C</code>, <code>G</code>, and <code>T</code> and &quot;any-mod&quot; as per the <a href="https://samtools.github.io/hts-specs/SAMtags.pdf">specification</a>.</p>
<h2 id="further-details"><a class="header" href="#further-details">Further details</a></h2>
<ol>
<li><a href="./filtering_details.html">Examples of how thresholds affect base modification calls.</a></li>
<li><a href="./filtering_numeric_details.html">Numerical details of how thresholds are calculated on the fly.</a></li>
</ol>
<div style="break-before: page; page-break-before: always;"></div><h1 id="examples-of-how-thresholds-affect-base-modification-calls"><a class="header" href="#examples-of-how-thresholds-affect-base-modification-calls">Examples of how thresholds affect base modification calls</a></h1>
<p>The following examples are meant to demonstrate how individual base modification calls will be made during
<code>pileup</code> or <code>call-mods</code>. Important to remember that the probability of the modification needs to be &gt;= the
pass threshold value.</p>
<h2 id="two-way-base-modification-calls"><a class="header" href="#two-way-base-modification-calls">Two-way base modification calls</a></h2>
<pre><code class="language-text">probability of canonical cytosine: p_C
probability of 5mC: p_m
threshold for all cytosine base modifications: threshold_C 

{p_C: 0.25, p_m: 0.75}, threshold_C: 0.7 =&gt; call: 5mC (m), most likely.
{p_C: 0.25, p_m: 0.75}, threshold_C: 0.8 =&gt; call: Fail, both probabilities are below threshold.
</code></pre>
<h2 id="three-way-base-modification-calls"><a class="header" href="#three-way-base-modification-calls">Three-way base modification calls</a></h2>
<pre><code class="language-text">probability of canonical cytosine: p_C
probability of 5mC: p_m
probability of 5hmC: p_h
threshold for all cytosine base modifications: threshold_C 

{p_C: 0.05, p_m: 0.7, p_h: 0.25}, threshold_C: 0.7 =&gt; call: 5mC (m), most likely.
{p_C: 0.15, p_m: 0.6, p_h: 0.25}, threshold_C: 0.7 =&gt; Fail, all below threshold.
</code></pre>
<h2 id="three-way-base-modification-calls-with-modification-specific-thresholds"><a class="header" href="#three-way-base-modification-calls-with-modification-specific-thresholds">Three-way base modification calls with modification-specific thresholds</a></h2>
<pre><code class="language-text">probability of canonical cytosine: p_C
probability of 5mC: p_m
probability of 5hmC: p_h
threshold for all canonical cytosine: mod_threshold_C 
threshold for all 5mC: mod_threshold_m 
threshold for all 5hmC: mod_threshold_h 

command line: --filter-threshold C:0.7 --mod-threshold m:0.8 --mod-threshold h:0.9

filter_threshold C: 0.7
mod_threshold_m: 0.8
mod_threshold_h: 0.9
{p_C: 0.05, p_m: 0.85, p_h: 0.1} =&gt; call: 5mC (m) most likely

filter_threshold C: 0.7
mod_threshold_m: 0.8
mod_threshold_h: 0.9
{p_C: 0.75, p_m: 0.05, p_h: 0.2} =&gt; call: C (canonical) most likely
{p_C: 0.05, p_m: 0.75, p_h: 0.2} =&gt; Fail, all below respective thresholds
{p_C: 0.1, p_m: 0.05, p_h: 0.85} =&gt; Fail, all below respective thresholds
</code></pre>
<div style="break-before: page; page-break-before: always;"></div><h1 id="filtering-base-modification-calls"><a class="header" href="#filtering-base-modification-calls">Filtering base modification calls.</a></h1>
<p>Modified base calls are qualitied with a probability that is contained in the ML tag (see the
<a href="https://samtools.github.io/hts-specs/SAMtags.pdf">specification</a>). We calculate the confidence that the model
has in the base modification prediction as \(\mathcal{q} = argmax(\textbf{P})\) where \(\textbf{P}\) is the
vector of probabilities for each modification. For example, given a model that can classify canonical
cytosine, 5mC, and 5hmC, \(\textbf{P}\) is \([P_{C}, P_m, P_h]\), and \(\mathcal{q}\) will be \(\mathcal{q} =
argmax(P_{C}, P_m, P_h)\), the maximum of the three probabilities.  Filtering in <code>modkit</code> is performed by
first determining the value of \(\mathcal{q}\) for the lowest n-th percentile of calls (10th percentile by
default).  The threshold value is typically an estimate because the base modification probabilities are
sampled from a subset of the reads. All calls can be used to calculate the exact value, but in practice the
approximation gives the same value. Base modification calls with a confidence value less than this number will
not be counted.  Determination of the threshold value can be performed on the fly (by sampling) or the
threshold value can be specified on the command line with the <code>--filter_threshold</code> flag. The <code>sample-probs</code>
command can be used to quickly estimate the value of \(\mathcal{q}\) at various percentiles.</p>
<h1 id="a-note-on-the-probability-of-modification--mm-flag"><a class="header" href="#a-note-on-the-probability-of-modification--mm-flag">A note on the &quot;probability of modification&quot; (<code>.</code>) MM flag.</a></h1>
<p>The <code>.</code> flag (e.g. <code>C+m.</code>) indicates that primary sequence bases without base modification calls can be inferred to be
canonical (<a href="https://samtools.github.io/hts-specs/SAMtags.pdf">SAM tags</a>). Some base modification callers,
for example <a href="https://github.com/nanoporetech/dorado/"><code>dorado</code></a> have a default threshold, below which a base modification
probability will be omitted (meaning it will be inferred to be a canonical/unmodified base). In general, omitting
the base modification probabilities when they are very confidently canonical will not change the results from <code>modkit</code>.
However, since the base modification probabilities are effectively changed to 100% canonical, can affect the
estimation of the pass threshold.</p>
<div style="break-before: page; page-break-before: always;"></div><h1 id="removing-modification-calls-from-bams"><a class="header" href="#removing-modification-calls-from-bams">Removing modification calls from BAMs.</a></h1>
<p>There may be multiple possible base modifications to a specific canonical base.
For example, cytosine has at least four known modified variants. As
technologies are increasingly able to detect these chemical moieties, not all
downstream tools may be capable of accepting them, or you may want to convert
your data to make it comparable to other data with less specific resolution. To
address this issue, <code>modkit</code> implements a method for removing one or more DNA
base modification call from a BAM as well as a command line flag to simply
combine all modification calls when performing a pileup to generate a bedMethyl
file.</p>
<h2 id="removing-dna-base-modification-probabilities"><a class="header" href="#removing-dna-base-modification-probabilities">Removing DNA base modification probabilities.</a></h2>
<p>BAM records containing probabilities for multiple base modifications at each residue can
be transformed to ignore one (or more) of the predicted base modifications. For example,
if a BAM file contains 5hmC and 5mC probabilities, the total probability is necessarily
distributed over 5mC, \(p_m\), 5hmC, \(p_h\), and canonical C, \(p_{C}\). <code>modkit</code>
implements a method, described below, for reducing the dimensionality of the probability
distribution by one or more of the predicted base modification classes.</p>
<p>Take for example a 3-way modification call for C, 5mC, and 5hmC, the probabilities of each
being \(p_{C}\), \(p_{m}\), \(p_{h}\), respectively.  To remove the 5hmC
probability, \(p_{h}\) is distributed evenly across the other two options. In this
example, the updates are \( p_C \leftarrow p_C + (\frac{p_h}{2}) \) and \( p_m
\leftarrow p_m + (\frac{p_h}{2}) \).</p>
<h2 id="combining-multiple-base-modifications-into-a-single-count"><a class="header" href="#combining-multiple-base-modifications-into-a-single-count">Combining multiple base modifications into a single count.</a></h2>
<h3 id="combine---combine"><a class="header" href="#combine---combine">Combine: <code>--combine</code></a></h3>
<p>In <code>modkit</code> the combine method can be used to produce binary modified/unmodified
counts. Continuing with the above example, the counts for number of modified reads,
<code>N_mod</code>, becomes the number of reads with 5hmC calls plus the number of reads with 5mC
calls. <code>N_other_mod</code> will always be <code>0</code>. The <code>raw_mod_code</code> will be reported as the
ambiguous mod code for the canonical base. See
https://samtools.github.io/hts-specs/SAMtags.pdf for details on base modification codes
and <code>schema.yaml</code> in this project for details on the notation for counts.</p>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->


                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">

            </nav>

        </div>




        <script>
            window.playground_copyable = true;
        </script>


        <script src="elasticlunr.min.js"></script>
        <script src="mark.min.js"></script>
        <script src="searcher.js"></script>

        <script src="clipboard.min.js"></script>
        <script src="highlight.js"></script>
        <script src="book.js"></script>

        <!-- Custom JS scripts -->

        <script>
        window.addEventListener('load', function() {
            MathJax.Hub.Register.StartupHook('End', function() {
                window.setTimeout(window.print, 100);
            });
        });
        </script>

    </div>
    </body>
</html>
