<!DOCTYPE HTML>
<html lang="en" class="sidebar-visible no-js light">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>Extended subcommand help - Modkit</title>


        <!-- Custom HTML head -->
        
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff" />

        <link rel="icon" href="favicon.svg">
        <link rel="shortcut icon" href="favicon.png">
        <link rel="stylesheet" href="css/variables.css">
        <link rel="stylesheet" href="css/general.css">
        <link rel="stylesheet" href="css/chrome.css">
        <link rel="stylesheet" href="css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="fonts/fonts.css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="highlight.css">
        <link rel="stylesheet" href="tomorrow-night.css">
        <link rel="stylesheet" href="ayu-highlight.css">

        <!-- Custom theme stylesheets -->
        <link rel="stylesheet" href="custom.css">

        <!-- MathJax -->
        <script async src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    </head>
    <body>
    <div id="body-container">
        <!-- Provide site root to javascript -->
        <script>
            var path_to_root = "";
            var default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? "navy" : "light";
        </script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script>
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script>
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            var html = document.querySelector('html');
            html.classList.remove('no-js')
            html.classList.remove('light')
            html.classList.add(theme);
            html.classList.add('js');
        </script>

        <!-- Hide / unhide sidebar before it is displayed -->
        <script>
            var html = document.querySelector('html');
            var sidebar = null;
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            } else {
                sidebar = 'hidden';
            }
            html.classList.remove('sidebar-visible');
            html.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <div class="sidebar-scrollbox">
                <ol class="chapter"><li class="chapter-item expanded "><a href="quick_start.html"><strong aria-hidden="true">1.</strong> Quick Start guides</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="intro_bedmethyl.html"><strong aria-hidden="true">1.1.</strong> Constructing bedMethyl tables</a></li><li class="chapter-item expanded "><a href="intro_adjust.html"><strong aria-hidden="true">1.2.</strong> Updating and adjusting MM tags</a></li><li class="chapter-item expanded "><a href="intro_summary.html"><strong aria-hidden="true">1.3.</strong> Summarizing a modBAM</a></li><li class="chapter-item expanded "><a href="intro_motif_bed.html"><strong aria-hidden="true">1.4.</strong> Making a motif BED file</a></li><li class="chapter-item expanded "><a href="intro_extract.html"><strong aria-hidden="true">1.5.</strong> Extracting read information to a table</a></li><li class="chapter-item expanded "><a href="intro_call_mods.html"><strong aria-hidden="true">1.6.</strong> Calling mods in a modBAM</a></li><li class="chapter-item expanded "><a href="intro_edge_filter.html"><strong aria-hidden="true">1.7.</strong> Removing modification calls at the ends of reads</a></li><li class="chapter-item expanded "><a href="intro_include_bed.html"><strong aria-hidden="true">1.8.</strong> Narrow output to specific positions</a></li><li class="chapter-item expanded "><a href="intro_repair.html"><strong aria-hidden="true">1.9.</strong> Repair MM/ML tags on trimmed reads</a></li><li class="chapter-item expanded "><a href="intro_pileup_hemi.html"><strong aria-hidden="true">1.10.</strong> Make hemi-methylation bedMethyl tables</a></li><li class="chapter-item expanded "><a href="intro_dmr.html"><strong aria-hidden="true">1.11.</strong> Perform differential methylation scoring</a></li></ol></li><li class="chapter-item expanded "><a href="advanced_usage.html" class="active"><strong aria-hidden="true">2.</strong> Extended subcommand help</a></li><li class="chapter-item expanded "><a href="troubleshooting.html"><strong aria-hidden="true">3.</strong> Troubleshooting</a></li><li class="chapter-item expanded "><a href="limitations.html"><strong aria-hidden="true">4.</strong> Current limitations</a></li><li class="chapter-item expanded "><a href="perf_considerations.html"><strong aria-hidden="true">5.</strong> Performance considerations</a></li><li class="chapter-item expanded "><a href="algo_details.html"><strong aria-hidden="true">6.</strong> Algorithm details</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="filtering.html"><strong aria-hidden="true">6.1.</strong> Pass/fail base modification calls</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="filtering_details.html"><strong aria-hidden="true">6.1.1.</strong> Threshold examples</a></li><li class="chapter-item expanded "><a href="filtering_numeric_details.html"><strong aria-hidden="true">6.1.2.</strong> Numeric details</a></li></ol></li><li class="chapter-item expanded "><a href="collapse.html"><strong aria-hidden="true">6.2.</strong> Ignoring and combining calls</a></li></ol></li></ol>
            </div>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle"></div>
        </nav>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky bordered">
                    <div class="left-buttons">
                        <button id="sidebar-toggle" class="icon-button" type="button" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </button>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="light">Light</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">Modkit</h1>

                    <div class="right-buttons">
                        <a href="print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>

                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script>
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <h1 id="modkit-subcommand-documentation"><a class="header" href="#modkit-subcommand-documentation">modkit, subcommand documentation</a></h1>
<p>The goal of <code>modkit</code> is to enable best-practices manipulation of BAM files containing
modified base information (modBAMs). The various sub-commands and tools available in
<code>modkit</code> are described below.  This information can be obtained by invoking the long help
(<code>--help</code>) for each command.</p>
<blockquote>
<p>Advanced usage information.</p>
</blockquote>
<pre><code class="language-text">Usage: modkit &lt;COMMAND&gt;

Commands:
  pileup        Tabulates base modification calls across genomic positions. This command
                    produces a bedMethyl formatted file. Schema and description of fields can be
                    found in the README.
  adjust-mods   Performs various operations on BAM files containing base modification
                    information, such as converting base modification codes and ignoring
                    modification calls. Produces a BAM output file.
  update-tags   Renames Mm/Ml to tags to MM/ML. Also allows changing the the mode flag from
                    silent '.' to explicitly '?' or '.'.
  sample-probs  Calculate an estimate of the base modification probability distribution.
  summary       Summarize the mod tags present in a BAM and get basic statistics. The default
                    output is a totals table (designated by '#' lines) and a modification calls
                    table. Descriptions of the columns can be found in the README.
  call-mods     Call mods from a modbam, creates a new modbam with probabilities set to 100% if
                    a base modification is called or 0% if called canonical.
  motif-bed     Create BED file with all locations of a sequence motif. 
                    Example: modkit motif-bed CG 0
  extract       Extract read-level base modification information from a modBAM into a
                    tab-separated values table.
  repair        Repair MM and ML tags in one bam with the correct tags from another. To use this
                    command, both modBAMs _must_ be sorted by read name. The &quot;donor&quot; modBAM's reads
                    must be a superset of the acceptor's reads. Extra reads in the donor are
                    allowed, and multiple reads with the same name (secondary, etc.) are allowed in
                    the acceptor. Reads with an empty SEQ field cannot be repaired and will be
                    rejected. Reads where there is an ambiguous alignment of the acceptor to the
                    donor will be rejected (and logged). See the full documentation for details.
  dmr           Perform DMR test on a set of regions. Output a BED file of regions with the
                    score column indicating the magnitude of the difference. Find the schema and
                    description of fields can in the README as well as a description of the model
                    and method. See subcommand help for additional details.
  pileup-hemi   Tabulates double-stranded base modification patters (such as hemi-methylation)
                    across genomic motif positions. This command produces a bedMethyl file, the
                    schema can be found in the online documentation.
  help          Print this message or the help of the given subcommand(s).

Options:
  -h, --help     Print help information.
  -V, --version  Print version information.
</code></pre>
<h2 id="pileup"><a class="header" href="#pileup">pileup</a></h2>
<pre><code class="language-text">Tabulates base modification calls across genomic positions. This command produces a bedMethyl
formatted file. Schema and description of fields can be found in the README.

Usage: modkit pileup [OPTIONS] &lt;IN_BAM&gt; &lt;OUT_BED&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input BAM, should be sorted and have associated index available.

  &lt;OUT_BED&gt;
          Output file (or directory with --bedgraph option) to write results into. Specify &quot;-&quot; or
          &quot;stdout&quot; to direct output to stdout.

Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them. Setting a file is
          recommended. (alias: log)

      --region &lt;REGION&gt;
          Process only the specified region of the BAM when performing pileup. Format should be
          &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;. Commas are allowed.

      --max-depth &lt;MAX_DEPTH&gt;
          Maximum number of records to use when calculating pileup. This argument is passed to the
          pileup engine. If you have high depth data, consider increasing this value substantially.
          Must be less than 2147483647 or an error will be raised.
          
          [default: 8000]

  -t, --threads &lt;THREADS&gt;
          Number of threads to use while processing chunks concurrently.
          
          [default: 4]

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently. Smaller interval chunk sizes
          will use less memory but incur more overhead.
          
          [default: 100000]

      --chunk-size &lt;CHUNK_SIZE&gt;
          Break contigs into chunks containing this many intervals (see `interval_size`). This
          option can be used to help prevent excessive memory usage, usually with no performance
          penalty. By default, modkit will set this value to 1.5x the number of threads specified,
          so if 4 threads are specified the chunk_size will be 6. A warning will be shown if this
          option is less than the number of threads specified.

      --suppress-progress
          Hide the progress bar.

  -n, --num-reads &lt;NUM_READS&gt;
          Sample this many reads when estimating the filtering threshold. Reads will be sampled
          evenly across aligned genome. If a region is specified, either with the --region option or
          the --sample-region option, then reads will be sampled evenly across the region given.
          This option is useful for large BAM files. In practice, 10-50 thousand reads is sufficient
          to estimate the model output distribution and determine the filtering threshold.
          
          [default: 10042]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Sample this fraction of the reads when estimating the pass-threshold. In practice, 10-100
          thousand reads is sufficient to estimate the model output distribution and determine the
          filtering threshold. See filtering.md for details on filtering.

      --seed &lt;SEED&gt;
          Set a random seed for deterministic running, the default is non-deterministic.

      --no-filtering
          Do not perform any filtering, include all mod base calls in output. See filtering.md for
          details on filtering.

  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted variant is below
          this confidence percentile. For example, 0.1 will filter out the 10% lowest confidence
          modification calls.
          
          [default: 0.1]

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or per-base. Global filter threshold can be
          specified with by a decimal number (e.g. 0.75). Per-base thresholds can be specified by
          colon-separated values, for example C:0.75 specifies a threshold value of 0.75 for
          cytosine modification calls. Additional per-base thresholds can be specified by repeating
          the option: for example --filter-threshold C:0.75 --filter-threshold A:0.70 or specify a
          single base option and a default for all other bases with: --filter-threshold A:0.70
          --filter-threshold 0.9 will specify a threshold value of 0.70 for adenine and 0.9 for all
          other base modification calls.

      --mod-thresholds &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification, independent of the threshold
          for the primary sequence base or the default. For example, to set the pass threshold for
          5hmC to 0.8 use `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless the
          `--filter-threshold` option is also passed. See the online documentation for more details.

      --sample-region &lt;SAMPLE_REGION&gt;
          Specify a region for sampling reads from when estimating the threshold probability. If
          this option is not provided, but --region is provided, the genomic interval passed to
          --region will be used. Format should be &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;.

      --sampling-interval-size &lt;SAMPLING_INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently when estimating the threshold
          probability, can be larger than the pileup processing interval.
          
          [default: 1000000]

      --include-bed &lt;INCLUDE_BED&gt;
          BED file that will restrict threshold estimation and pileup results to positions
          overlapping intervals in the file. (alias: include-positions)

      --include-unmapped
          Include unmapped base modifications when estimating the pass threshold.

      --ignore &lt;IGNORE&gt;
          Ignore a modified base class  _in_situ_ by redistributing base modification probability
          equally across other options. For example, if collapsing 'h', with 'm' and canonical
          options, half of the probability of 'h' will be added to both 'm' and 'C'. A full
          description of the methods can be found in collapse.md.

      --force-allow-implicit
          Force allow implicit-canonical mode. By default modkit does not allow pileup with the
          implicit mode (e.g. C+m, no '.' or '?'). The `update-tags` subcommand is provided to
          update tags to the new mode. This option allows the interpretation of implicit mode tags:
          residues without modified base probability will be interpreted as being the non-modified
          base.

      --motif &lt;MOTIF&gt; &lt;MOTIF&gt;
          Output pileup counts for only sequence motifs provided. The first argument should be the
          sequence motif and the second argument is the 0-based offset to the base to pileup base
          modification counts for. For example: --motif CGCG 0 indicates to pileup counts for the
          first C on the top strand and the last C (complement to G) on the bottom strand. The --cpg
          argument is short hand for --motif CG 0.
          
          This argument can be passed multiple times. When more than one motif is used, the
          resulting output BED file will indicate the motif in the &quot;name&quot; field as
          &lt;mod_code&gt;,&lt;motif&gt;,&lt;offset&gt;. For example, given `--motif CGCG 2 --motif CG 0` there will
          be output lines with name fields such as &quot;m,CG,0&quot; and &quot;m,CGCG,2&quot;. To use this option with
          `--combine-strands`, all motifs must be reverse-complement palindromic or an error will be
          raised.

      --cpg
          Only output counts at CpG motifs. Requires a reference sequence to be provided.

  -r, --ref &lt;REFERENCE_FASTA&gt;
          Reference sequence in FASTA format. Required for CpG motif filtering.

  -k, --mask
          Respect soft masking in the reference FASTA.

      --preset &lt;PRESET&gt;
          Optional preset options for specific applications. traditional: Prepares bedMethyl
          analogous to that generated from other technologies for the analysis of 5mC modified
          bases. Shorthand for --cpg --combine-strands --ignore h.
          
          [possible values: traditional]

      --combine-mods
          Combine base modification calls, all counts of modified bases are summed together. See
          collapse.md for details.

      --combine-strands
          When performing motif analysis (such as CpG), sum the counts from the positive and
          negative strands into the counts for the positive strand position.

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the start or the end of the
          read. Two comma-separated values may be provided to asymmetrically filter out base
          modification calls from the start and end of the reads. For example, 4,8 will filter out
          base modification calls in the first 4 and last 8 bases of the read.

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification calls at the ends of
          reads, only _keep_ base modification calls at the ends of reads. E.g. if usually, &quot;4,8&quot;
          would remove (i.e. filter out) base modification calls in the first 4 and last 8 bases of
          the read, using this flag will keep only base modification calls in the first 4 and last 8
          bases.

      --only-tabs
          For bedMethyl output, separate columns with only tabs. The default is to use tabs for the
          first 10 fields and spaces thereafter. The default behavior is more likely to be
          compatible with genome viewers. Enabling this option may make it easier to parse the
          output with tabular data handlers that expect a single kind of separator.

      --bedgraph
          Output bedGraph format, see https://genome.ucsc.edu/goldenPath/help/bedgraph.html. For
          this setting, specify a directory for output files to be make in. Two files for each
          modification will be produced, one for the positive strand and one for the negative
          strand. So for 5mC (m) and 5hmC (h) there will be 4 files produced.

      --prefix &lt;PREFIX&gt;
          Prefix to prepend on bedgraph output file names. Without this option the files will be
          &lt;mod_code&gt;_&lt;strand&gt;.bedgraph.

      --partition-tag &lt;PARTITION_TAG&gt;
          Partition output into multiple bedMethyl files based on tag-value pairs. The output will
          be multiple bedMethyl files with the format
          `&lt;prefix&gt;_&lt;tag_value_1&gt;_&lt;tag_value_2&gt;_&lt;tag_value_n&gt;.bed` prefix is optional and set with
          the `--prefix` flag.

  -h, --help
          Print help information (use `-h` for a summary)
</code></pre>
<h2 id="adjust-mods"><a class="header" href="#adjust-mods">adjust-mods</a></h2>
<pre><code class="language-text">Performs various operations on BAM files containing base modification information, such as
converting base modification codes and ignoring modification calls. Produces a BAM output file

Usage: modkit adjust-mods [OPTIONS] &lt;IN_BAM&gt; &lt;OUT_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          BAM file to collapse mod call from. Can be a path to a file or one of `-` or `stdin` to
          specify a stream from standard input.

  &lt;OUT_BAM&gt;
          File path to new BAM file to be created. Can be a path to a file or one of `-` or `stdin`
          to specify a stream from standard output.

Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Output debug logs to file at this path.

      --ignore &lt;IGNORE&gt;
          Modified base code to ignore/remove, see https://samtools.github.io/hts-specs/SAMtags.pdf
          for details on the modified base codes.

  -t, --threads &lt;THREADS&gt;
          Number of threads to use.
          
          [default: 4]

  -f, --ff
          Fast fail, stop processing at the first invalid sequence record. Default behavior is to
          continue and report failed/skipped records at the end.

      --convert &lt;CONVERT&gt; &lt;CONVERT&gt;
          Convert one mod-tag to another, summing the probabilities together if the retained mod tag
          is already present.

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the start or the end of the
          read. Two comma-separated values may be provided to asymmetrically filter out base
          modification calls from the start and end of the reads. For example, 4,8 will filter out
          base modification calls in the first 4 and last 8 bases of the read.

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification calls at the ends of
          reads, only _keep_ base modification calls at the ends of reads. E.g. if usually, &quot;4,8&quot;
          would remove (i.e. filter out) base modification calls in the first 4 and last 8 bases of
          the read, using this flag will keep only base modification calls in the first 4 and last 8
          bases.

      --output-sam
          Output SAM format instead of BAM.

      --suppress-progress
          Hide the progress bar

  -h, --help
          Print help information (use `-h` for a summary).
</code></pre>
<h2 id="update-tags"><a class="header" href="#update-tags">update-tags</a></h2>
<pre><code class="language-text">Renames Mm/Ml to tags to MM/ML. Also allows changing the the mode flag from silent '.' to explicitly
'?' or '.'.

Usage: modkit update-tags [OPTIONS] &lt;IN_BAM&gt; &lt;OUT_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;   BAM to update modified base tags in. Can be a path to a file or one of `-` or `stdin`
             to specify a stream from standard input.
  &lt;OUT_BAM&gt;  File to new BAM file to be created or one of `-` or `stdin` to specify a stream from
             standard output.

Options:
  -m, --mode &lt;MODE&gt;                  Mode, change mode to this value, options {'ambiguous',
                                     'implicit'}. See spec at:
                                     https://samtools.github.io/hts-specs/SAMtags.pdf. 'ambiguous'
                                     ('?') means residues without explicit modification
                                     probabilities will not be assumed canonical or modified.
                                     'implicit' means residues without explicit modification
                                     probabilities are assumed to be canonical. [possible values:
                                     ambiguous, implicit]
  -t, --threads &lt;THREADS&gt;            Number of threads to use. [default: 4]
      --log-filepath &lt;LOG_FILEPATH&gt;  Output debug logs to file at this path.
      --output-sam                   Output SAM format instead of BAM.
  -h, --help                         Print help information.
</code></pre>
<h2 id="sample-probs"><a class="header" href="#sample-probs">sample-probs</a></h2>
<pre><code class="language-text">Calculate an estimate of the base modification probability distribution.

Usage: modkit sample-probs [OPTIONS] &lt;IN_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input BAM with modified base tags. If a index is found reads will be sampled evenly across
          the length of the reference sequence. Can be a path to a file or one of `-` or `stdin` to
          specify a stream from standard input.

Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use.
          
          [default: 4]

      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them. Setting a file is
          recommended.

      --suppress-progress
          Hide the progress bar.

  -p, --percentiles &lt;PERCENTILES&gt;
          Percentiles to calculate, a space separated list of floats.
          
          [default: 0.1,0.5,0.9]

  -o, --out-dir &lt;OUT_DIR&gt;
          Directory to deposit result tables into. Required for model probability histogram output.
          Creates two files probabilities.tsv and probabilities.txt The .txt contains
          ASCII-histograms and the .tsv contains tab-separated variable data represented by the
          histograms.

      --prefix &lt;PREFIX&gt;
          Label to prefix output files with. E.g. 'foo' will output foo_thresholds.tsv,
          foo_probabilities.tsv, and foo_probabilities.txt.

      --force
          Overwrite results if present.

      --ignore &lt;IGNORE&gt;
          Ignore a modified base class  _in_situ_ by redistributing base modification probability
          equally across other options. For example, if collapsing 'h', with 'm' and canonical
          options, half of the probability of 'h' will be added to both 'm' and 'C'. A full
          description of the methods can be found in collapse.md.

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the start or the end of the
          read. Two comma-separated values may be provided to asymmetrically filter out base
          modification calls from the start and end of the reads. For example, 4,8 will filter out
          base modification calls in the first 4 and last 8 bases of the read.

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification calls at the ends of
          reads, only _keep_ base modification calls at the ends of reads. E.g. if usually, &quot;4,8&quot;
          would remove (i.e. filter out) base modification calls in the first 4 and last 8 bases of
          the read, using this flag will keep only base modification calls in the first 4 and last 8
          bases.

      --hist
          Output histogram of base modification prediction probabilities.

      --buckets &lt;BUCKETS&gt;
          Number of buckets for the histogram, if used.
          
          [default: 128]

  -n, --num-reads &lt;NUM_READS&gt;
          Approximate maximum number of reads to use, especially recommended when using a large BAM
          without an index. If an indexed BAM is provided, the reads will be sampled evenly over the
          length of the aligned reference. If a region is passed with the --region option, they will
          be sampled over the genomic region. Actual number of reads used may deviate slightly from
          this number.
          
          [default: 10042]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Instead of using a defined number of reads, specify a fraction of reads to sample, for
          example 0.1 will sample 1/10th of the reads.

      --no-sampling
          No sampling, use all of the reads to calculate the filter thresholds.

  -s, --seed &lt;SEED&gt;
          Random seed for deterministic running, the default is non-deterministic, only used when no
          BAM index is provided.

      --region &lt;REGION&gt;
          Process only the specified region of the BAM when collecting probabilities. Format should
          be &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;.

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently. Smaller interval chunk sizes
          will use less memory but incur more overhead. Only used when sampling probs from an
          indexed bam.
          
          [default: 1000000]

      --include-bed &lt;INCLUDE_BED&gt;
          Only sample base modification probabilities that are aligned to the positions in this BED
          file. (alias: include-positions)

      --only-mapped
          Only use base modification probabilities that are aligned (i.e. ignore soft-clipped, and
          inserted bases).

  -h, --help
          Print help information (use `-h` for a summary).
</code></pre>
<h2 id="summary"><a class="header" href="#summary">summary</a></h2>
<pre><code class="language-text">Summarize the mod tags present in a BAM and get basic statistics. The default output is a totals
table (designated by '#' lines) and a modification calls table. Descriptions of the columns can be
found in the README.

Usage: modkit summary [OPTIONS] &lt;IN_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input modBam, can be a path to a file or one of `-` or `stdin` to specify a stream from
          standard input.

Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use.
          
          [default: 4]

      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them. Setting a file is
          recommended.

      --tsv
          Output summary as a tab-separated variables stdout instead of a table.

      --suppress-progress
          Hide the progress bar.

  -n, --num-reads &lt;NUM_READS&gt;
          Approximate maximum number of reads to use, especially recommended when using a large BAM
          without an index. If an indexed BAM is provided, the reads will be sampled evenly over the
          length of the aligned reference. If a region is passed with the --region option, they will
          be sampled over the genomic region. Actual number of reads used may deviate slightly from
          this number.
          
          [default: 10042]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Instead of using a defined number of reads, specify a fraction of reads to sample when
          estimating the filter threshold. For example 0.1 will sample 1/10th of the reads.

      --no-sampling
          No sampling, use all of the reads to calculate the filter thresholds and generating the
          summary.

  -s, --seed &lt;SEED&gt;
          Sets a random seed for deterministic running (when using --sample-frac), the default is
          non-deterministic, only used when no BAM index is provided.

      --no-filtering
          Do not perform any filtering, include all base modification calls in the summary. See
          filtering.md for details on filtering.

  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted variant is below
          this confidence percentile. For example, 0.1 will filter out the 10% lowest confidence
          base modification calls.
          
          [default: 0.1]

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or per-base. Global filter threshold can be
          specified with by a decimal number (e.g. 0.75). Per-base thresholds can be specified by
          colon-separated values, for example C:0.75 specifies a threshold value of 0.75 for
          cytosine modification calls. Additional per-base thresholds can be specified by repeating
          the option: for example --filter-threshold C:0.75 --filter-threshold A:0.70 or specify a
          single base option and a default for all other bases with: --filter-threshold A:0.70
          --filter-threshold 0.9 will specify a threshold value of 0.70 for adenine and 0.9 for all
          other base modification calls.

      --mod-thresholds &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification, independent of the threshold
          for the primary sequence base or the default. For example, to set the pass threshold for
          5hmC to 0.8 use `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless the
          `--filter-threshold` option is also passed. See the online documentation for more details.

      --ignore &lt;IGNORE&gt;
          Ignore a modified base class  _in_situ_ by redistributing base modification probability
          equally across other options. For example, if collapsing 'h', with 'm' and canonical
          options, half of the probability of 'h' will be added to both 'm' and 'C'. A full
          description of the methods can be found in collapse.md.

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the start or the end of the
          read. Two comma-separated values may be provided to asymmetrically filter out base
          modification calls from the start and end of the reads. For example, 4,8 will filter out
          base modification calls in the first 4 and last 8 bases of the read.

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification calls at the ends of
          reads, only _keep_ base modification calls at the ends of reads. E.g. if usually, &quot;4,8&quot;
          would remove (i.e. filter out) base modification calls in the first 4 and last 8 bases of
          the read, using this flag will keep only base modification calls in the first 4 and last 8
          bases.

      --include-bed &lt;INCLUDE_BED&gt;
          Only summarize base modification probabilities that are aligned to the positions in this
          BED file. (alias: include-positions)

      --only-mapped
          Only use base modification probabilities that are aligned (i.e. ignore soft-clipped, and
          inserted bases).

      --region &lt;REGION&gt;
          Process only the specified region of the BAM when collecting probabilities. Format should
          be &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;.

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          When using regions, interval chunk size in base pairs to process concurrently. Smaller
          interval chunk sizes will use less memory but incur more overhead.
          
          [default: 1000000]

  -h, --help
          Print help information (use `-h` for a summary).
</code></pre>
<h2 id="motif-bed"><a class="header" href="#motif-bed">motif-bed</a></h2>
<pre><code class="language-text">Create BED file with all locations of a sequence motif. Example: modkit motif-bed CG 0

Usage: modkit motif-bed [OPTIONS] &lt;FASTA&gt; &lt;MOTIF&gt; &lt;OFFSET&gt;

Arguments:
  &lt;FASTA&gt;   Input FASTA file.
  &lt;MOTIF&gt;   Motif to search for within FASTA, e.g. CG.
  &lt;OFFSET&gt;  Offset within motif, e.g. 0.

Options:
  -k, --mask  Respect soft masking in the reference FASTA.
  -h, --help  Print help information.
</code></pre>
<h2 id="call-mods"><a class="header" href="#call-mods">call-mods</a></h2>
<pre><code class="language-text">Call mods from a modBam, creates a new modBam with probabilities set to 100% if a base modification
is called or 0% if called canonical.

Usage: modkit call-mods [OPTIONS] &lt;IN_BAM&gt; &lt;OUT_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input BAM, may be sorted and have associated index available. Can be a path to a file or
          one of `-` or `stdin` to specify a stream from standard input.

  &lt;OUT_BAM&gt;
          Output BAM, can be a path to a file or one of `-` or `stdin` to specify a stream from
          standard input.

Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them. Setting a file is
          recommended.

      --ff
          Fast fail, stop processing at the first invalid sequence record. Default behavior is to
          continue and report failed/skipped records at the end.

      --suppress-progress
          Hide the progress bar.

  -t, --threads &lt;THREADS&gt;
          Number of threads to use while processing chunks concurrently.
          
          [default: 4]

  -n, --num-reads &lt;NUM_READS&gt;
          Sample approximately this many reads when estimating the filtering threshold. If
          alignments are present reads will be sampled evenly across aligned genome. If a region is
          specified, either with the --region option or the --sample-region option, then reads will
          be sampled evenly across the region given. This option is useful for large BAM files. In
          practice, 10-50 thousand reads is sufficient to estimate the model output distribution and
          determine the filtering threshold.
          
          [default: 10042]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Sample this fraction of the reads when estimating the filter-percentile. In practice,
          50-100 thousand reads is sufficient to estimate the model output distribution and
          determine the filtering threshold. See filtering.md for details on filtering.

      --seed &lt;SEED&gt;
          Set a random seed for deterministic running, the default is non-deterministic, only used
          when no BAM index is provided.

      --sample-region &lt;SAMPLE_REGION&gt;
          Specify a region for sampling reads from when estimating the threshold probability. If
          this option is not provided, but --region is provided, the genomic interval passed to
          --region will be used. Format should be &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;.

      --sampling-interval-size &lt;SAMPLING_INTERVAL_SIZE&gt;
          Interval chunk size to process concurrently when estimating the threshold probability, can
          be larger than the pileup processing interval.
          
          [default: 1000000]

  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted variant is below
          this confidence percentile. For example, 0.1 will filter out the 10% lowest confidence
          modification calls.
          
          [default: 0.1]

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or per primary base. A global filter threshold can
          be specified with by a decimal number (e.g. 0.75). Per-base thresholds can be specified by
          colon-separated values, for example C:0.75 specifies a threshold value of 0.75 for
          cytosine modification calls. Additional per-base thresholds can be specified by repeating
          the option: for example --filter-threshold C:0.75 --filter-threshold A:0.70 or specify a
          single base option and a default for all other bases with: --filter-threshold A:0.70
          --filter-threshold 0.9 will specify a threshold value of 0.70 for adenine and 0.9 for all
          other base modification calls.

      --mod-threshold &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification, independent of the threshold
          for the primary sequence base or the default. For example, to set the pass threshold for
          5hmC to 0.8 use `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless the
          `--filter-threshold` option is also passed. See the online documentation for more details.

      --no-filtering
          Don't filter base modification calls, assign each base modification to the highest
          probability prediction.

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the start or the end of the
          read. Two comma-separated values may be provided to asymmetrically filter out base
          modification calls from the start and end of the reads. For example, 4,8 will filter out
          base modification calls in the first 4 and last 8 bases of the read.

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification calls at the ends of
          reads, only _keep_ base modification calls at the ends of reads. E.g. if usually, &quot;4,8&quot;
          would remove (i.e. filter out) base modification calls in the first 4 and last 8 bases of
          the read, using this flag will keep only base modification calls in the first 4 and last 8
          bases.

      --output-sam
          Output SAM format instead of BAM.

  -h, --help
          Print help information (use `-h` for a summary).
</code></pre>
<h2 id="extract"><a class="header" href="#extract">extract</a></h2>
<pre><code class="language-text">Extract read-level base modification information from a modBAM into a tab-separated values table.

Usage: modkit extract [OPTIONS] &lt;IN_BAM&gt; &lt;OUT_PATH&gt;

Arguments:
  &lt;IN_BAM&gt;
          Path to modBAM file to extract read-level information from, or one of `-` or `stdin` to
          specify a stream from standard input. If a file is used it may be sorted and have
          associated index.

  &lt;OUT_PATH&gt;
          Path to output file, &quot;stdout&quot; or &quot;-&quot; will direct output to standard out. Specifying &quot;null&quot;
          will not output the extract table (useful if all you need is the `--read-calls` output
          table).

Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use.
          
          [default: 4]

      --log-filepath &lt;LOG_FILEPATH&gt;
          Path to file to write run log, setting this file is recommended.

      --mapped-only
          Include only mapped bases in output. (alias: mapped)

      --allow-non-primary
          Output aligned secondary and supplementary base modification probabilities as additional
          rows. The primary alignment will have all of the base modification probabilities
          (including soft-clipped ones, unless --mapped-only is used). The non-primary alignments
          will only have mapped bases in the output.

      --num-reads &lt;NUM_READS&gt;
          Number of reads to use. Note that when using a sorted, indexed modBAM that the sampling
          algorithm will attempt to sample records evenly over the length of the reference sequence.
          The result is the final number of records used may be slightly more or less than the
          requested number. When piping from stdin or using a modBAM without an index, the requested
          number of reads will be exact.

      --region &lt;REGION&gt;
          Process only reads that are aligned to a specified region of the BAM. Format should be
          &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;.

      --force
          Force overwrite of output file.

      --suppress-progress
          Hide the progress bar.

      --kmer-size &lt;KMER_SIZE&gt;
          Set the query and reference k-mer size (if a reference is provided). Maxumum number for
          this value is 12.
          
          [default: 5]

      --ignore-index
          Ignore the BAM index (if it exists) and default to a serial scan of the BAM.

      --read-calls-path &lt;READ_CALLS_PATH&gt;
          Produce a table of read-level base modification calls. This table has, for each read, one
          row for each base modification call in that read using the same thresholding algorithm as
          in pileup, or summary (see online documentation for details on thresholds). Passing this
          option will cause `modkit` to estimate the pass thresholds from the data unless a
          `--filter-threshold` value is passed to the command. (alias: --read-calls)

      --reference &lt;REFERENCE&gt;
          Path to reference FASTA to extract reference context information from. If no reference is
          provided, `ref_kmer` column will be &quot;.&quot; in the output. (alias: ref)

      --include-bed &lt;INCLUDE_BED&gt;
          BED file with regions to include (alias: include-positions). Implicitly only includes
          mapped sites

  -v, --exclude-bed &lt;EXCLUDE_BED&gt;
          BED file with regions to _exclude_ (alias: exclude).

      --motif &lt;MOTIF&gt; &lt;MOTIF&gt;
          Output read-level base modification probabilities restricted to the reference sequence
          motifs provided. The first argument should be the sequence motif and the second argument
          is the 0-based offset to the base to pileup base modification counts for. For example:
          --motif CGCG 0 indicates include base modifications for which the read is aligned to the
          first C on the top strand and the last C (complement to G) on the bottom strand. The --cpg
          argument is short hand for --motif CG 0. This argument can be passed multiple times.

      --cpg
          Only output counts at CpG motifs. Requires a reference sequence to be provided.

  -k, --mask
          When using motifs, respect soft masking in the reference sequence.

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or per-base. Global filter threshold can be
          specified with by a decimal number (e.g. 0.75). Per-base thresholds can be specified by
          colon-separated values, for example C:0.75 specifies a threshold value of 0.75 for
          cytosine modification calls. Additional per-base thresholds can be specified by repeating
          the option: for example --filter-threshold C:0.75 --filter-threshold A:0.70 or specify a
          single base option and a default for all other bases with: --filter-threshold A:0.70
          --filter-threshold 0.9 will specify a threshold value of 0.70 for adenine and 0.9 for all
          other base modification calls.

      --mod-thresholds &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification, independent of the threshold
          for the primary sequence base or the default. For example, to set the pass threshold for
          5hmC to 0.8 use `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless the
          `--filter-threshold` option is also passed. See the online documentation for more details.

      --no-filtering
          Do not perform any filtering, include all mod base calls in output. See filtering.md for
          details on filtering.

      --sampling-interval-size &lt;SAMPLING_INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently when estimating the threshold
          probability.
          
          [default: 1000000]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Sample this fraction of the reads when estimating the pass-threshold. In practice, 10-100
          thousand reads is sufficient to estimate the model output distribution and determine the
          filtering threshold. See filtering.md for details on filtering.

  -n, --sample-num-reads &lt;SAMPLE_NUM_READS&gt;
          Sample this many reads when estimating the filtering threshold. If a sorted, indexed
          modBAM is provided reads will be sampled evenly across aligned genome. If a region is
          specified, with the --region, then reads will be sampled evenly across the region given.
          This option is useful for large BAM files. In practice, 10-50 thousand reads is sufficient
          to estimate the model output distribution and determine the filtering threshold.
          
          [default: 10042]

      --seed &lt;SEED&gt;
          Set a random seed for deterministic running, the default is non-deterministic.

  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted variant is below
          this confidence percentile. For example, 0.1 will filter out the 10% lowest confidence
          modification calls.
          
          [default: 0.1]

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the start or the end of the
          read. Two comma-separated values may be provided to asymmetrically filter out base
          modification calls from the start and end of the reads. For example, 4,8 will filter out
          base modification calls in the first 4 and last 8 bases of the read.

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification calls at the ends of
          reads, only _keep_ base modification calls at the ends of reads. E.g. if usually, &quot;4,8&quot;
          would remove (i.e. filter out) base modification calls in the first 4 and last 8 bases of
          the read, using this flag will keep only base modification calls in the first 4 and last 8
          bases.

      --ignore &lt;IGNORE&gt;
          Ignore a modified base class  _in_situ_ by redistributing base modification probability
          equally across other options. For example, if collapsing 'h', with 'm' and canonical
          options, half of the probability of 'h' will be added to both 'm' and 'C'. A full
          description of the methods can be found in collapse.md.

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently. Smaller interval chunk sizes
          will use less memory but incur more overhead. Only used when an indexed modBam is provided.
          
          [default: 100000]

      --ignore-implicit
          Ignore implicitly canonical base modification calls. When the `.` flag is used in the MM
          tag, this implies that bases missing a base modification probability are to be assumed
          canonical. Set this flag to omit those base modifications from the output. For additional
          details see the SAM spec: https://samtools.github.io/hts-specs/SAMtags.pdf.

  -h, --help
          Print help information (use `-h` for a summary).
</code></pre>
<h2 id="repair"><a class="header" href="#repair">repair</a></h2>
<pre><code class="language-text">Repair MM and ML tags in one bam with the correct tags from another. To use this command, both
modBAMs _must_ be sorted by read name. The &quot;donor&quot; modBAM's reads must be a superset of the
acceptor's reads. Extra reads in the donor are allowed, and multiple reads with the same name
(secondary, etc.) are allowed in the acceptor. Reads with an empty SEQ field cannot be repaired and
will be rejected. Reads where there is an ambiguous alignment of the acceptor to the donor will be
rejected (and logged). See the full documentation for details.

Usage: modkit repair [OPTIONS] --donor-bam &lt;DONOR_BAM&gt; --acceptor-bam &lt;ACCEPTOR_BAM&gt; --output-bam &lt;OUTPUT_BAM&gt;

Options:
  -d, --donor-bam &lt;DONOR_BAM&gt;        Donor modBAM with original MM/ML tags. Must be sorted by read
                                     name.
  -a, --acceptor-bam &lt;ACCEPTOR_BAM&gt;  Acceptor modBAM with reads to have MM/ML base modification data
                                     projected on to. Must be sorted by read name.
  -o, --output-bam &lt;OUTPUT_BAM&gt;      output modBAM location.
      --log-filepath &lt;LOG_FILEPATH&gt;  File to write logs to, it is recommended to use this option as
                                     some reads may be rejected and logged here
  -t, --threads &lt;THREADS&gt;            The number of threads to use [default: 4]
  -h, --help                         Print help information.
</code></pre>
<h2 id="pileup-hemi"><a class="header" href="#pileup-hemi">pileup-hemi</a></h2>
<pre><code class="language-text">Tabulates double-stranded base modification patters (such as hemi-methylation) across genomic motif
positions. This command produces a bedMethyl file, the schema can be found in the online
documentation.

Usage: modkit pileup-hemi [OPTIONS] &lt;IN_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input BAM, should be sorted and have associated index available.

Options:
  -o, --out-bed &lt;OUT_BED&gt;
          Output file to write results into. Will write to stdout if not provided.

      --cpg
          Aggregate double-stranded base modifications for CpG dinucleotides. This flag is
          short-hand for --motif CG 0.

      --motif &lt;MOTIF&gt; &lt;MOTIF&gt;
          Specify the sequence motif to pileup double-stranded base modification pattern counts for.
          The first argument should be the sequence motif and the second argument is the 0-based
          offset to the base to pileup base modification counts for. For example: --motif CG 0
          indicates to generate pattern counts for the C on the top strand and the following C
          (opposite to G) on the negative strand. The motif must be reverse-complement palindromic
          or an error will be raised. See the documentation for more examples and details.

  -r, --ref &lt;REFERENCE_FASTA&gt;
          Reference sequence in FASTA format.

      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them. Setting a file is
          recommended. (alias: log)

      --region &lt;REGION&gt;
          Process only the specified region of the BAM when performing pileup. Format should be
          &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;. Commas are allowed.

      --max-depth &lt;MAX_DEPTH&gt;
          Maximum number of records to use when calculating pileup. This argument is passed to the
          pileup engine. If you have high depth data, consider increasing this value substantially.
          Must be less than 2147483647 or an error will be raised.
          
          [default: 8000]

  -t, --threads &lt;THREADS&gt;
          Number of threads to use while processing chunks concurrently.
          
          [default: 4]

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently. Smaller interval chunk sizes
          will use less memory but incur more overhead.
          
          [default: 100000]

      --chunk-size &lt;CHUNK_SIZE&gt;
          Break contigs into chunks containing this many intervals (see `interval_size`). This
          option can be used to help prevent excessive memory usage, usually with no performance
          penalty. By default, modkit will set this value to 1.5x the number of threads specified,
          so if 4 threads are specified the chunk_size will be 6. A warning will be shown if this
          option is less than the number of threads specified.

      --suppress-progress
          Hide the progress bar.

  -n, --num-reads &lt;NUM_READS&gt;
          Sample this many reads when estimating the filtering threshold. Reads will be sampled
          evenly across aligned genome. If a region is specified, either with the --region option or
          the --sample-region option, then reads will be sampled evenly across the region given.
          This option is useful for large BAM files. In practice, 10-50 thousand reads is sufficient
          to estimate the model output distribution and determine the filtering threshold.
          
          [default: 10042]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Sample this fraction of the reads when estimating the filter-percentile. In practice,
          50-100 thousand reads is sufficient to estimate the model output distribution and
          determine the filtering threshold. See filtering.md for details on filtering.

      --seed &lt;SEED&gt;
          Set a random seed for deterministic running, the default is non-deterministic.

      --no-filtering
          Do not perform any filtering, include all mod base calls in output. See filtering.md for
          details on filtering.

  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted variant is below
          this confidence percentile. For example, 0.1 will filter out the 10% lowest confidence
          modification calls.
          
          [default: 0.1]

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or per-base. Global filter threshold can be
          specified with by a decimal number (e.g. 0.75). Per-base thresholds can be specified by
          colon-separated values, for example C:0.75 specifies a threshold value of 0.75 for
          cytosine modification calls. Additional per-base thresholds can be specified by repeating
          the option: for example --filter-threshold C:0.75 --filter-threshold A:0.70 or specify a
          single base option and a default for all other bases with: --filter-threshold A:0.70
          --filter-threshold 0.9 will specify a threshold value of 0.70 for adenine and 0.9 for all
          other base modification calls.

      --mod-thresholds &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification, independent of the threshold
          for the primary sequence base or the default. For example, to set the pass threshold for
          5hmC to 0.8 use `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless the
          `--filter-threshold` option is also passed. See the online documentation for more details.

      --sample-region &lt;SAMPLE_REGION&gt;
          Specify a region for sampling reads from when estimating the threshold probability. If
          this option is not provided, but --region is provided, the genomic interval passed to
          --region will be used. Format should be &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;.

      --sampling-interval-size &lt;SAMPLING_INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently when estimating the threshold
          probability, can be larger than the pileup processing interval.
          
          [default: 1000000]

      --include-bed &lt;INCLUDE_BED&gt;
          BED file that will restrict threshold estimation and pileup results to positions
          overlapping intervals in the file. (alias: include-positions)

      --include-unmapped
          Include unmapped base modifications when estimating the pass threshold.

      --ignore &lt;IGNORE&gt;
          Ignore a modified base class  _in_situ_ by redistributing base modification probability
          equally across other options. For example, if collapsing 'h', with 'm' and canonical
          options, half of the probability of 'h' will be added to both 'm' and 'C'. A full
          description of the methods can be found in collapse.md.

      --force-allow-implicit
          Force allow implicit-canonical mode. By default modkit does not allow pileup with the
          implicit mode (e.g. C+m, no '.' or '?'). The `update-tags` subcommand is provided to
          update tags to the new mode. This option allows the interpretation of implicit mode tags:
          residues without modified base probability will be interpreted as being the non-modified
          base.

  -k, --mask
          Respect soft masking in the reference FASTA.

      --combine-mods
          Combine base modification calls, all counts of modified bases are summed together. See
          collapse.md for details.

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the start or the end of the
          read. Two comma-separated values may be provided to asymmetrically filter out base
          modification calls from the start and end of the reads. For example, 4,8 will filter out
          base modification calls in the first 4 and last 8 bases of the read.

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification calls at the ends of
          reads, only _keep_ base modification calls at the ends of reads. E.g. if usually, &quot;4,8&quot;
          would remove (i.e. filter out) base modification calls in the first 4 and last 8 bases of
          the read, using this flag will keep only base modification calls in the first 4 and last 8
          bases.

      --only-tabs
          Separate bedMethyl columns with only tabs. The default is to use tabs for the first 10
          fields and spaces thereafter. The default behavior is more likely to be compatible with
          genome viewers. Enabling this option may make it easier to parse the output with tabular
          data handlers that expect a single kind of separator.

  -h, --help
          Print help information (use `-h` for a summary).
</code></pre>
<h2 id="pileup-hemi-pair"><a class="header" href="#pileup-hemi-pair">pileup-hemi <code>pair</code></a></h2>
<pre><code class="language-text">Compare regions in a pair of samples (for example, tumor and normal or control and experiment). A
sample is input as a bgzip pileup bedMethyl (produced by pileup, for example) that has an associated
tabix index. Output is a BED file with the score column indicating the magnitude of the difference
in methylation between the two samples. See the online documentation for additional details.

Usage: modkit dmr pair [OPTIONS] -a &lt;CONTROL_BED_METHYL&gt; -b &lt;EXP_BED_METHYL&gt; --ref &lt;REFERENCE_FASTA&gt;

Options:
  -a &lt;CONTROL_BED_METHYL&gt;
          Bgzipped bedMethyl file for the first (usually control) sample. There should be a tabix
          index with the same name and .tbi next to this file or the --index-a option must be
          provided.
  -b &lt;EXP_BED_METHYL&gt;
          Bgzipped bedMethyl file for the second (usually experimental) sample. There should be a
          tabix index with the same name and .tbi next to this file or the --index-b option must be
          provided.
  -o, --out-path &lt;OUT_PATH&gt;
          Path to file to direct output, optional, no argument will direct output to stdout.
  -r, --regions-bed &lt;REGIONS_BED&gt;
          BED file of regions over which to compare methylation levels. Should be tab-separated
          (spaces allowed in the &quot;name&quot; column). Requires chrom, chromStart and chromEnd. The Name
          column is optional. Strand is currently ignored. When omitted, methylation levels are
          compared at each site in the `-a`/`control_bed_methyl` BED file (or optionally, the
          `-b`/`exp_bed_methyl` file with the `--use-b` flag.
      --use-b
          When performing site-level DMR, use the bedMethyl indicated by the -b/exp_bed_methyl
          argument to collect bases to score.
      --ref &lt;REFERENCE_FASTA&gt;
          Path to reference fasta for the pileup.
  -m &lt;MODIFIED_BASES&gt;
          Bases to use to calculate DMR, may be multiple. For example, to calculate differentially
          methylated regions using only cytosine modifications use --base C.
      --log-filepath &lt;LOG_FILEPATH&gt;
          File to write logs to, it's recommended to use this option.
  -t, --threads &lt;THREADS&gt;
          Number of threads to use [default: 4]
      --batch-size &lt;BATCH_SIZE&gt;
          Control the  batch size. The batch size is the number of regions to load at a time. Each
          region will be processed concurrently. Loading more regions at a time will decrease IO to
          load data, but will use more memory. Default will be 50% more than the number of threads
          assigned.
  -k, --mask
          Respect soft masking in the reference FASTA.
      --suppress-progress
          Don't show progress bars.
  -f, --force
          Force overwrite of output file, if it already exists.
      --index-a &lt;INDEX_A&gt;
          Path to tabix index associated with -a (--control-bed-methyl) bedMethyl file.
      --index-b &lt;INDEX_B&gt;
          Path to tabix index associated with -b (--exp-bed-methyl) bedMethyl file.
      --missing &lt;HANDLE_MISSING&gt;
          How to handle regions found in the `--regions` BED file. quiet =&gt; ignore regions that are
          not found in the tabix header warn =&gt; log (debug) regions that are missing fatal =&gt; log
          (error) and exit the program when a region is missing. [default: warn] [possible values:
          quiet, warn, fail]
      --min-valid-coverage &lt;MIN_VALID_COVERAGE&gt;
          Minimum valid coverage required to use an entry from a bedMethyl. See the help for pileup
          for the specification and description of valid coverage. [default: 0]
  -h, --help
          Print help information.
</code></pre>
<h2 id="pileup-hemi-1"><a class="header" href="#pileup-hemi-1">pileup-hemi</a></h2>
<pre><code class="language-text">Compare regions between all pairs of samples (for example a trio sample set or haplotyped trio
sample set). As with `pair` all inputs must be bgzip compressed bedMethyl files with associated
tabix indices. Each sample must be assigned a name. Output is a directory of BED files with the
score column indicating the magnitude of the difference in methylation between the two samples
indicated in the file name. See the online documentation for additional details.

Usage: modkit dmr multi [OPTIONS] --out-dir &lt;OUT_DIR&gt; --ref &lt;REFERENCE_FASTA&gt;

Options:
  -s, --sample &lt;SAMPLES&gt; &lt;SAMPLES&gt;
          Two or more named samples to compare. Two arguments are required &lt;path&gt; &lt;name&gt;. This
          option should be repeated at least two times.
  -i, --index &lt;INDICES&gt; &lt;INDICES&gt;
          Optional, paths to tabix indices associated with named samples. Two arguments are required
          &lt;path&gt; &lt;name&gt; where &lt;name&gt; corresponds to the name of the sample given to the -s/--sample
          argument.
  -r, --regions-bed &lt;REGIONS_BED&gt;
          BED file of regions over which to compare methylation levels. Should be tab-separated
          (spaces allowed in the &quot;name&quot; column). Requires chrom, chromStart and chromEnd. The Name
          column is optional. Strand is currently ignored. When omitted, methylation levels are
          compared at each site in common between the two bedMethyl files being compared.
  -o, --out-dir &lt;OUT_DIR&gt;
          Directory to place output DMR results in BED format.
  -p, --prefix &lt;PREFIX&gt;
          Prefix files in directory with this label.
      --ref &lt;REFERENCE_FASTA&gt;
          Path to reference fasta for the pileup.
  -m &lt;MODIFIED_BASES&gt;
          Bases to use to calculate DMR, may be multiple. For example, to calculate differentially
          methylated regions using only cytosine modifications use --base C.
      --log-filepath &lt;LOG_FILEPATH&gt;
          File to write logs to, it's recommended to use this option.
  -t, --threads &lt;THREADS&gt;
          Number of threads to use. [default: 4]
  -k, --mask
          Respect soft masking in the reference FASTA.
      --suppress-progress
          Don't show progress bars.
  -f, --force
          Force overwrite of output file, if it already exists
      --missing &lt;HANDLE_MISSING&gt;
          How to handle regions found in the `--regions` BED file. quiet =&gt; ignore regions that are
          not found in the tabix header warn =&gt; log (debug) regions that are missing fatal =&gt; log
          (error) and exit the program when a region is missing. [default: warn] [possible values:
          quiet, warn, fail]
      --min-valid-coverage &lt;MIN_VALID_COVERAGE&gt;
          Minimum valid coverage required to use an entry from a bedMethyl. See the help for pileup
          for the specification and description of valid coverage. [default: 0]
  -h, --help
          Print help information
</code></pre>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->
                            <a rel="prev" href="intro_dmr.html" class="mobile-nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                                <i class="fa fa-angle-left"></i>
                            </a>

                            <a rel="next" href="troubleshooting.html" class="mobile-nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                                <i class="fa fa-angle-right"></i>
                            </a>

                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">
                    <a rel="prev" href="intro_dmr.html" class="nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                        <i class="fa fa-angle-left"></i>
                    </a>

                    <a rel="next" href="troubleshooting.html" class="nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                        <i class="fa fa-angle-right"></i>
                    </a>
            </nav>

        </div>




        <script>
            window.playground_copyable = true;
        </script>


        <script src="elasticlunr.min.js"></script>
        <script src="mark.min.js"></script>
        <script src="searcher.js"></script>

        <script src="clipboard.min.js"></script>
        <script src="highlight.js"></script>
        <script src="book.js"></script>

        <!-- Custom JS scripts -->


    </div>
    </body>
</html>
