<!DOCTYPE HTML>
<html lang="en" class="sidebar-visible no-js light">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>Constructing bedMethyl tables - Modkit</title>


        <!-- Custom HTML head -->
        
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff" />

        <link rel="icon" href="favicon.svg">
        <link rel="shortcut icon" href="favicon.png">
        <link rel="stylesheet" href="css/variables.css">
        <link rel="stylesheet" href="css/general.css">
        <link rel="stylesheet" href="css/chrome.css">
        <link rel="stylesheet" href="css/print.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="FontAwesome/css/font-awesome.css">
        <link rel="stylesheet" href="fonts/fonts.css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" href="highlight.css">
        <link rel="stylesheet" href="tomorrow-night.css">
        <link rel="stylesheet" href="ayu-highlight.css">

        <!-- Custom theme stylesheets -->
        <link rel="stylesheet" href="custom.css">

        <!-- MathJax -->
        <script async src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    </head>
    <body>
    <div id="body-container">
        <!-- Provide site root to javascript -->
        <script>
            var path_to_root = "";
            var default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? "navy" : "light";
        </script>

        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script>
            try {
                var theme = localStorage.getItem('mdbook-theme');
                var sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script>
            var theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            var html = document.querySelector('html');
            html.classList.remove('no-js')
            html.classList.remove('light')
            html.classList.add(theme);
            html.classList.add('js');
        </script>

        <!-- Hide / unhide sidebar before it is displayed -->
        <script>
            var html = document.querySelector('html');
            var sidebar = null;
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            } else {
                sidebar = 'hidden';
            }
            html.classList.remove('sidebar-visible');
            html.classList.add("sidebar-" + sidebar);
        </script>

        <nav id="sidebar" class="sidebar" aria-label="Table of contents">
            <div class="sidebar-scrollbox">
                <ol class="chapter"><li class="chapter-item expanded "><a href="quick_start.html"><strong aria-hidden="true">1.</strong> Quick Start guides</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="intro_bedmethyl.html" class="active"><strong aria-hidden="true">1.1.</strong> Constructing bedMethyl tables</a></li><li class="chapter-item expanded "><a href="intro_adjust.html"><strong aria-hidden="true">1.2.</strong> Updating and adjusting MM tags</a></li><li class="chapter-item expanded "><a href="intro_summary.html"><strong aria-hidden="true">1.3.</strong> Summarizing a modBAM</a></li><li class="chapter-item expanded "><a href="intro_motif_bed.html"><strong aria-hidden="true">1.4.</strong> Making a motif BED file</a></li><li class="chapter-item expanded "><a href="intro_extract.html"><strong aria-hidden="true">1.5.</strong> Extracting read information to a table</a></li><li class="chapter-item expanded "><a href="intro_call_mods.html"><strong aria-hidden="true">1.6.</strong> Calling mods in a modBAM</a></li><li class="chapter-item expanded "><a href="intro_edge_filter.html"><strong aria-hidden="true">1.7.</strong> Removing modification calls at the ends of reads</a></li><li class="chapter-item expanded "><a href="intro_include_bed.html"><strong aria-hidden="true">1.8.</strong> Narrow output to specific positions</a></li><li class="chapter-item expanded "><a href="intro_repair.html"><strong aria-hidden="true">1.9.</strong> Repair MM/ML tags on trimmed reads</a></li><li class="chapter-item expanded "><a href="intro_pileup_hemi.html"><strong aria-hidden="true">1.10.</strong> Make hemi-methylation bedMethyl tables</a></li><li class="chapter-item expanded "><a href="intro_dmr.html"><strong aria-hidden="true">1.11.</strong> Perform differential methylation scoring</a></li></ol></li><li class="chapter-item expanded "><a href="advanced_usage.html"><strong aria-hidden="true">2.</strong> Extended subcommand help</a></li><li class="chapter-item expanded "><a href="troubleshooting.html"><strong aria-hidden="true">3.</strong> Troubleshooting</a></li><li class="chapter-item expanded "><a href="limitations.html"><strong aria-hidden="true">4.</strong> Current limitations</a></li><li class="chapter-item expanded "><a href="perf_considerations.html"><strong aria-hidden="true">5.</strong> Performance considerations</a></li><li class="chapter-item expanded "><a href="algo_details.html"><strong aria-hidden="true">6.</strong> Algorithm details</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="filtering.html"><strong aria-hidden="true">6.1.</strong> Pass/fail base modification calls</a></li><li><ol class="section"><li class="chapter-item expanded "><a href="filtering_details.html"><strong aria-hidden="true">6.1.1.</strong> Threshold examples</a></li><li class="chapter-item expanded "><a href="filtering_numeric_details.html"><strong aria-hidden="true">6.1.2.</strong> Numeric details</a></li></ol></li><li class="chapter-item expanded "><a href="collapse.html"><strong aria-hidden="true">6.2.</strong> Ignoring and combining calls</a></li></ol></li></ol>
            </div>
            <div id="sidebar-resize-handle" class="sidebar-resize-handle"></div>
        </nav>

        <div id="page-wrapper" class="page-wrapper">

            <div class="page">
                                <div id="menu-bar-hover-placeholder"></div>
                <div id="menu-bar" class="menu-bar sticky bordered">
                    <div class="left-buttons">
                        <button id="sidebar-toggle" class="icon-button" type="button" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="sidebar">
                            <i class="fa fa-bars"></i>
                        </button>
                        <button id="theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="theme-list">
                            <i class="fa fa-paint-brush"></i>
                        </button>
                        <ul id="theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="light">Light</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="ayu">Ayu</button></li>
                        </ul>
                        <button id="search-toggle" class="icon-button" type="button" title="Search. (Shortkey: s)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="S" aria-controls="searchbar">
                            <i class="fa fa-search"></i>
                        </button>
                    </div>

                    <h1 class="menu-title">Modkit</h1>

                    <div class="right-buttons">
                        <a href="print.html" title="Print this book" aria-label="Print this book">
                            <i id="print-button" class="fa fa-print"></i>
                        </a>

                    </div>
                </div>

                <div id="search-wrapper" class="hidden">
                    <form id="searchbar-outer" class="searchbar-outer">
                        <input type="search" id="searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="searchresults-outer" aria-describedby="searchresults-header">
                    </form>
                    <div id="searchresults-outer" class="searchresults-outer hidden">
                        <div id="searchresults-header" class="searchresults-header"></div>
                        <ul id="searchresults">
                        </ul>
                    </div>
                </div>

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script>
                    document.getElementById('sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="content" class="content">
                    <main>
                        <h1 id="constructing-bedmethyl-tables"><a class="header" href="#constructing-bedmethyl-tables">Constructing bedMethyl tables.</a></h1>
<p>A primary use of <code>modkit</code> is to create summary counts of modified and unmodified bases in
an extended <a href="https://www.encodeproject.org/data-standards/wgbs/">bedMethyl</a> format.
bedMethyl files tabulate the counts of base modifications from every sequencing read over
each aligned reference genomic position. In order to create a bedMethyl table, your modBAM
must be aligned to a reference genome. The genome sequence is only required if you are using
the <code>--cpg</code> flag or <code>traditional</code> preset. Only <strong>primary alignments</strong> are used in generating 
the table, it is recommended to mark duplicate alignments before running as multiple primary
alignments can be double counted (but the behavior is logged). See <a href="./limitations.html">limitations</a>
for details.</p>
<h2 id="basic-usage"><a class="header" href="#basic-usage">Basic usage</a></h2>
<p>In its simplest form <code>modkit</code> creates a bedMethyl file using the following:</p>
<pre><code class="language-text">modkit pileup path/to/reads.bam output/path/pileup.bed --log-filepath pileup.log
</code></pre>
<p>No reference sequence is required. A single file (described
<a href="#description-of-bedmethyl-output">below</a>) with base count summaries will be created. The
final argument here specifies an optional log file output.</p>
<p>The program performs best-practices filtering and manipulation of the raw data stored in
the input file. For further details see <a href="./filtering.html">filtering modified-base calls</a>.</p>
<h3 id="narrowing-output-to-cpg-dinucleotides"><a class="header" href="#narrowing-output-to-cpg-dinucleotides">Narrowing output to CpG dinucleotides</a></h3>
<p>For user convenience, the counting process can be modulated using several additional
transforms and filters. The most basic of these is to report only counts from reference
CpG dinucleotides. This option requires a reference sequence in order to locate the CpGs
in the reference:</p>
<pre><code class="language-bash">modkit pileup path/to/reads.bam output/path/pileup.bed --cpg --ref path/to/reference.fasta
</code></pre>
<p>To restrict output to only certain CpGs, pass the <code>--include-bed</code> option with the CpGs to be used, 
see <a href="./intro_include_bed.html">this page</a> for more details.</p>
<pre><code class="language-bash">modkit pileup path/to/reads.bam output/path/pileup.bed \
  --cpg \
  --ref path/to/reference.fasta \
  --include-bed path/to/my_cpgs.bed
</code></pre>
<p>The program also contains preset which combine several options for ease of use. The
<code>traditional</code> preset,</p>
<pre><code class="language-bash">modkit pileup path/to/reads.bam output/path/pileup.bed \
  --ref path/to/reference.fasta \
  --preset traditional
</code></pre>
<p>performs three transforms:</p>
<ul>
<li>restricts output to locations where there is a CG dinucleotide in the reference,</li>
<li>reports only a C and 5mC counts, using procedures to take into account counts of other
forms of cytosine modification (notably 5hmC), and</li>
<li>aggregates data across strands. The strand field of the output will be marked as '.'
indicating that the strand information has been lost.</li>
</ul>
<p>Using this option is equivalent to running with the options:</p>
<pre><code class="language-bash">modkit pileup path/to/reads.bam output/path/pileup.bed --cpg --ref &lt;reference.fasta&gt; --ignore h --combine-strands
</code></pre>
<h3 id="narrowing-output-to-specific-motifs"><a class="header" href="#narrowing-output-to-specific-motifs">Narrowing output to specific motifs</a></h3>
<p>By default, <code>modkit</code> will output a BED row for all genomic positions where
there is at least one base modification in the input modBAM. We define a motif
as a short DNA sequence potentially containing <a href="https://en.wikipedia.org/wiki/Nucleic_acid_notation">degenerate
codes</a>. To ease downstream
analysis, the <code>--motif &lt;Motif&gt; &lt;offset, 0-based&gt;</code> option can be used to
pre-filter and annotate the bedMethyl rows. The <code>--cpg</code> flag is a alias for
<code>--motif CG 0</code> where the sequence motif is <code>CG</code> and the offset is <code>0</code>, meaning
pileup base modification counts for the first <code>C</code> in the motif on the top
strand the second <code>C</code> (complement to <code>G</code>) on the bottom strand. Another example
may be <code>--motif GATC 1</code>, signaling to pileup counts for the <code>A</code> in the second
position on the top strand and the <code>A</code> in the third position on the bottom
strand.</p>
<p>When multiple motifs are specified the <code>name</code> column (<a href="#bedmethyl-column-descriptions">column
4</a>), will indicate which motif the counts are
tabulated for. For example, if <code>--motif CGCG 2 --motif CG 0</code> are passed you may
see lines such as:</p>
<pre><code class="language-text">oligo_741_adapters  39 40 m,CG,0   4	-	39	40	255,0,0	4 100.00 4 0 0 0 0 0 0
oligo_741_adapters  39 40 m,CGCG,2 4	-	39	40	255,0,0	4 100.00 4 0 0 0 0 0 0

</code></pre>
<p>The <code>--combine-strands</code> flag can be combined with <code>--motif</code> however all motifs
must be reverse-complement palindromic (<code>CG</code> <em>is</em> a palindrome but <code>CHH</code> is
not).</p>
<h3 id="partitioning-reads-based-on-sam-tag-values"><a class="header" href="#partitioning-reads-based-on-sam-tag-values">Partitioning reads based on SAM tag values</a></h3>
<p>If have a modBAM with reads from different conditions are other SAM tag annotations (for example <code>RG</code> or <code>HP</code>) you 
can pass the <code>--partition-tag</code> option and <code>modkit</code> will output a separate bedMethyl with counts for only the reads 
with that tag value. For example, if you have haplotype-annotated reads with the <code>HP</code> tag, you could use a command
like the following:</p>
<pre><code class="language-bash">modkit pileup path/to/reads.bam output/directory/ --cpg --ref &lt;reference.fasta&gt; --partition-tag HP --prefix haplotyped
</code></pre>
<p>The output will be multiple files in placed in <code>output/directory/haplotyped_&lt;1|2|etc&gt;.bed</code>, multiple <code>--partition-tag</code>
options can be passed and the output files will correspond to the observed combinations of tags found in the modBAM. 
For example if <code>--partition-tag RG</code> and <code>--partition-tag HP</code> are passed:</p>
<pre><code class="language-bash">outdir/
  &lt;prefix&gt;_&lt;RG_value_1&gt;_&lt;HP_value_1&gt;.bed
  &lt;prefix&gt;_&lt;RG_value_2&gt;_&lt;HP_value_1&gt;.bed
  &lt;prefix&gt;_&lt;RG_value_1&gt;_&lt;HP_value_2&gt;.bed
  &lt;prefix&gt;_&lt;RG_value_2&gt;_&lt;HP_value_2&gt;.bed
  # ... etc
</code></pre>
<p>Note that only tag values that can be easily turned into strings will be considered valid (e.g. numbers, characters,
strings, etc.), array values will not be used, and will result in <code>missing</code> being used. Reads missing all of the 
SAM tags will be put in <code>ungrouped.bed</code>.</p>
<p>For more information on the individual options see the <a href="./advanced_usage.html">Advanced Usage</a> help document.</p>
<h2 id="description-of-bedmethyl-output"><a class="header" href="#description-of-bedmethyl-output">Description of bedMethyl output.</a></h2>
<p>Below is a description of the bedMethyl columns generated by <code>modkit pileup</code>. A brief description of the
bedMethyl specification can be found on <a href="https://www.encodeproject.org/data-standards/wgbs/">Encode</a>.</p>
<h3 id="definitions"><a class="header" href="#definitions">Definitions:</a></h3>
<ul>
<li>N<sub>mod</sub> - Number of calls passing filters that were classified as a residue with a specified base modification.</li>
<li>N<sub>canonical</sub> - Number of calls passing filters were classified as the canonical base rather than modified. The
exact base must be inferred by the modification code. For example, if the modification code is <code>m</code> (5mC) then
the canonical base is cytosine. If the modification code is <code>a</code>, the canonical base is adenine.</li>
<li>N<sub>other mod</sub> - Number of calls passing filters that were classified as modified, but where the modification is different from the listed base (and the corresponding canonical base is equal). For example, for a given cytosine there may be 3 reads with
<code>h</code> calls, 1 with a canonical call, and 2 with <code>m</code> calls. In the bedMethyl row for <code>h</code> N<sub>other_mod</sub> would be 2. In the
<code>m</code> row N<sub>other_mod</sub> would be 3.</li>
<li>N<sub>valid_cov</sub> - the valid coverage. N<sub>valid_cov</sub> = N<sub>mod</sub> + N<sub>other_mod</sub> + N<sub>canonical</sub>, also used as the <code>score</code> in the bedMethyl</li>
<li>N<sub>diff</sub> - Number of reads with a base other than the canonical base for this modification. For example, in a row
for <code>h</code> the canonical base is cytosine, if there are 2 reads with C-&gt;A substitutions, N<sub>diff</sub> will be 2.</li>
<li>N<sub>delete</sub> - Number of reads with a deletion at this reference position</li>
<li>N<sub>fail</sub> - Number of calls where the probability of the call was below the threshold. The threshold can be
set on the command line or computed from the data (usually failing the lowest 10th percentile of calls).</li>
<li>N<sub>nocall</sub> - Number of reads aligned to this reference position, with the correct canonical base, but without a base
modification call. This can happen, for example, if the model requires a CpG dinucleotide and the read has a
CG-&gt;CH substitution such that no modification call was produced by the basecaller.</li>
</ul>
<h3 id="bedmethyl-column-descriptions"><a class="header" href="#bedmethyl-column-descriptions">bedMethyl column descriptions.</a></h3>
<div class="table-wrapper"><table><thead><tr><th>column</th><th>name</th><th>description</th><th>type</th></tr></thead><tbody>
<tr><td>1</td><td>chrom</td><td>name of reference sequence from BAM header</td><td>str</td></tr>
<tr><td>2</td><td>start position</td><td>0-based start position</td><td>int</td></tr>
<tr><td>3</td><td>end position</td><td>0-based exclusive end position</td><td>int</td></tr>
<tr><td>4</td><td>modified base code and motif</td><td>single letter code for modified base and motif when more than one motif is used</td><td>str</td></tr>
<tr><td>5</td><td>score</td><td>Equal to N<sub>valid_cov</sub>.</td><td>int</td></tr>
<tr><td>6</td><td>strand</td><td>'+' for positive strand '-' for negative strand, '.' when strands are combined</td><td>str</td></tr>
<tr><td>7</td><td>start position</td><td>included for compatibility</td><td>int</td></tr>
<tr><td>8</td><td>end position</td><td>included for compatibility</td><td>int</td></tr>
<tr><td>9</td><td>color</td><td>included for compatibility, always 255,0,0</td><td>str</td></tr>
<tr><td>10</td><td>N<sub>valid_cov</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>11</td><td>fraction modified</td><td>N<sub>mod</sub> / N<sub>valid_cov</sub></td><td>float</td></tr>
<tr><td>12</td><td>N<sub>mod</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>13</td><td>N<sub>canonical</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>14</td><td>N<sub>other_mod</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>15</td><td>N<sub>delete</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>16</td><td>N<sub>fail</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>17</td><td>N<sub>diff</sub></td><td>See definitions above.</td><td>int</td></tr>
<tr><td>18</td><td>N<sub>nocall</sub></td><td>See definitions above.</td><td>int</td></tr>
</tbody></table>
</div>
<h2 id="performance-considerations"><a class="header" href="#performance-considerations">Performance considerations</a></h2>
<p>The <code>--interval-size</code>, <code>--threads</code>, <code>--chunk-size</code>, and <code>--max-depth</code> parameters can be used to tweak the parallelism and 
memory consumption of <code>modkit pileup</code>. The defaults should be suitable for most use cases, for more details see the
<a href="./advanced_usage.html">advanced usage</a> and <a href="./perf_considerations.html">performance considerations</a> sections.</p>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->
                            <a rel="prev" href="quick_start.html" class="mobile-nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                                <i class="fa fa-angle-left"></i>
                            </a>

                            <a rel="next" href="intro_adjust.html" class="mobile-nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                                <i class="fa fa-angle-right"></i>
                            </a>

                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">
                    <a rel="prev" href="quick_start.html" class="nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                        <i class="fa fa-angle-left"></i>
                    </a>

                    <a rel="next" href="intro_adjust.html" class="nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                        <i class="fa fa-angle-right"></i>
                    </a>
            </nav>

        </div>




        <script>
            window.playground_copyable = true;
        </script>


        <script src="elasticlunr.min.js"></script>
        <script src="mark.min.js"></script>
        <script src="searcher.js"></script>

        <script src="clipboard.min.js"></script>
        <script src="highlight.js"></script>
        <script src="book.js"></script>

        <!-- Custom JS scripts -->


    </div>
    </body>
</html>
